// @ng-dynamic-forms/core/esm2015/lib/model/form-group/dynamic-form-group.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DynamicFormControlModel } from "../dynamic-form-control.model";
import { serializable } from "../../decorator/serializable.decorator";
/** @type {?} */
export const DYNAMIC_FORM_CONTROL_TYPE_GROUP = "GROUP";
/**
 * @record
 */
export function DynamicFormGroupModelConfig() { }
if (false) {
    /** @type {?|undefined} */
    DynamicFormGroupModelConfig.prototype.group;
    /** @type {?|undefined} */
    DynamicFormGroupModelConfig.prototype.legend;
}
export class DynamicFormGroupModel extends DynamicFormControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.group = [];
        this.type = DYNAMIC_FORM_CONTROL_TYPE_GROUP;
        this.group = Array.isArray(config.group) ? config.group : [];
        this.legend = config.legend || null;
    }
    /**
     * @param {?} index
     * @return {?}
     */
    get(index) {
        return this.group[index];
    }
    /**
     * @param {?} index
     * @param {?} controlModel
     * @return {?}
     */
    set(index, controlModel) {
        this.group[index] = controlModel;
    }
    /**
     * @param {?} controlModel
     * @return {?}
     */
    add(controlModel) {
        this.group.push(controlModel);
    }
    /**
     * @param {?} index
     * @param {?} controlModel
     * @return {?}
     */
    insert(index, controlModel) {
        this.group.splice(index, 0, controlModel);
    }
    /**
     * @param {?} index
     * @param {?} step
     * @return {?}
     */
    move(index, step) {
        this.group.splice(index + step, 0, ...this.group.splice(index, 1));
    }
    /**
     * @param {?} index
     * @return {?}
     */
    remove(index) {
        this.group.splice(index, 1);
    }
    /**
     * @return {?}
     */
    size() {
        return this.group.length;
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Array)
], DynamicFormGroupModel.prototype, "group", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicFormGroupModel.prototype, "legend", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicFormGroupModel.prototype, "type", void 0);
if (false) {
    /** @type {?} */
    DynamicFormGroupModel.prototype.group;
    /** @type {?} */
    DynamicFormGroupModel.prototype.legend;
    /** @type {?} */
    DynamicFormGroupModel.prototype.type;
}
