// @ng-dynamic-forms/core/esm5/lib/model/checkbox/dynamic-checkbox-group.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DynamicFormGroupModel } from "../form-group/dynamic-form-group.model";
import { serializable } from "../../decorator/serializable.decorator";
/** @type {?} */
export var DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP = "CHECKBOX_GROUP";
var DynamicCheckboxGroupModel = /** @class */ (function (_super) {
    tslib_1.__extends(DynamicCheckboxGroupModel, _super);
    function DynamicCheckboxGroupModel(config, layout) {
        var _this = _super.call(this, config, layout) || this;
        _this.type = DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP;
        return _this;
    }
    /**
     * @param {...?} indices
     * @return {?}
     */
    DynamicCheckboxGroupModel.prototype.check = /**
     * @param {...?} indices
     * @return {?}
     */
    function () {
        var _this = this;
        var indices = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            indices[_i] = arguments[_i];
        }
        indices.forEach((/**
         * @param {?} index
         * @return {?}
         */
        function (index) { return _this.group[index].checked = true; }));
    };
    /**
     * @param {...?} indices
     * @return {?}
     */
    DynamicCheckboxGroupModel.prototype.uncheck = /**
     * @param {...?} indices
     * @return {?}
     */
    function () {
        var _this = this;
        var indices = [];
        for (var _i = 0; _i < arguments.length; _i++) {
            indices[_i] = arguments[_i];
        }
        indices.forEach((/**
         * @param {?} index
         * @return {?}
         */
        function (index) { return _this.group[index].checked = false; }));
    };
    /**
     * @return {?}
     */
    DynamicCheckboxGroupModel.prototype.checkAll = /**
     * @return {?}
     */
    function () {
        this.group.forEach((/**
         * @param {?} model
         * @return {?}
         */
        function (model) { return model.checked = true; }));
    };
    /**
     * @return {?}
     */
    DynamicCheckboxGroupModel.prototype.uncheckAll = /**
     * @return {?}
     */
    function () {
        this.group.forEach((/**
         * @param {?} model
         * @return {?}
         */
        function (model) { return model.checked = false; }));
    };
    tslib_1.__decorate([
        serializable(),
        tslib_1.__metadata("design:type", Array)
    ], DynamicCheckboxGroupModel.prototype, "group", void 0);
    tslib_1.__decorate([
        serializable(),
        tslib_1.__metadata("design:type", String)
    ], DynamicCheckboxGroupModel.prototype, "type", void 0);
    return DynamicCheckboxGroupModel;
}(DynamicFormGroupModel));
export { DynamicCheckboxGroupModel };
if (false) {
    /** @type {?} */
    DynamicCheckboxGroupModel.prototype.group;
    /** @type {?} */
    DynamicCheckboxGroupModel.prototype.type;
}
