// @ng-dynamic-forms/core/esm2015/lib/model/radio/dynamic-radio-group.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DynamicOptionControlModel } from "../dynamic-option-control.model";
import { serializable } from "../../decorator/serializable.decorator";
/** @type {?} */
export const DYNAMIC_FORM_CONTROL_TYPE_RADIO_GROUP = "RADIO_GROUP";
/**
 * @record
 * @template T
 */
export function DynamicRadioGroupModelConfig() { }
if (false) {
    /** @type {?|undefined} */
    DynamicRadioGroupModelConfig.prototype.legend;
}
/**
 * @template T
 */
export class DynamicRadioGroupModel extends DynamicOptionControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_RADIO_GROUP;
        this.legend = config.legend || null;
    }
    /**
     * @param {?} index
     * @return {?}
     */
    select(index) {
        this.value = this.get(index).value;
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicRadioGroupModel.prototype, "legend", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicRadioGroupModel.prototype, "type", void 0);
if (false) {
    /** @type {?} */
    DynamicRadioGroupModel.prototype.legend;
    /** @type {?} */
    DynamicRadioGroupModel.prototype.type;
}
