// @ng-dynamic-forms/core/esm2015/lib/model/textarea/dynamic-textarea.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DynamicInputControlModel } from "../dynamic-input-control.model";
import { serializable } from "../../decorator/serializable.decorator";
import { isNumber } from "../../utils/core.utils";
/** @type {?} */
export const DYNAMIC_FORM_CONTROL_TYPE_TEXTAREA = "TEXTAREA";
/** @type {?} */
export const DYNAMIC_FORM_TEXTAREA_WRAP_HARD = "hard";
/** @type {?} */
export const DYNAMIC_FORM_TEXTAREA_WRAP_SOFT = "soft";
/**
 * @record
 */
export function DynamicTextAreaModelConfig() { }
if (false) {
    /** @type {?|undefined} */
    DynamicTextAreaModelConfig.prototype.cols;
    /** @type {?|undefined} */
    DynamicTextAreaModelConfig.prototype.rows;
    /** @type {?|undefined} */
    DynamicTextAreaModelConfig.prototype.wrap;
}
export class DynamicTextAreaModel extends DynamicInputControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_TEXTAREA;
        this.cols = isNumber(config.cols) ? config.cols : 20;
        this.rows = isNumber(config.rows) ? config.rows : 2;
        this.wrap = config.wrap || DYNAMIC_FORM_TEXTAREA_WRAP_SOFT;
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Number)
], DynamicTextAreaModel.prototype, "cols", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Number)
], DynamicTextAreaModel.prototype, "rows", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicTextAreaModel.prototype, "wrap", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicTextAreaModel.prototype, "type", void 0);
if (false) {
    /** @type {?} */
    DynamicTextAreaModel.prototype.cols;
    /** @type {?} */
    DynamicTextAreaModel.prototype.rows;
    /** @type {?} */
    DynamicTextAreaModel.prototype.wrap;
    /** @type {?} */
    DynamicTextAreaModel.prototype.type;
}
