// @ng-dynamic-forms/core/fesm2015/core.js
import { __decorate, __metadata } from 'tslib';
import { BehaviorSubject, isObservable, of, merge } from 'rxjs';
import 'reflect-metadata';
import { tap, map, startWith } from 'rxjs/operators';
import { Directive, ElementRef, Renderer2, Input, TemplateRef, ɵlooseIdentical, InjectionToken, Injectable, Optional, Inject, ɵɵdefineInjectable, ɵɵinject, QueryList, Injector, INJECTOR, NgModule } from '@angular/core';
import { Validators, NG_VALIDATORS, NG_ASYNC_VALIDATORS, FormArray, FormControl, FormGroup, ReactiveFormsModule } from '@angular/forms';
import { CommonModule } from '@angular/common';

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} value
 * @return {?}
 */
function isBoolean(value) {
    return typeof value === "boolean";
}
/**
 * @param {?} value
 * @return {?}
 */
function isFunction(value) {
    return typeof value === "function";
}
/**
 * @param {?} value
 * @return {?}
 */
function isNumber(value) {
    return typeof value === "number";
}
/**
 * @param {?} value
 * @return {?}
 */
function isObject(value) {
    return typeof value === "object" && value !== null;
}
/**
 * @param {?} value
 * @return {?}
 */
function isString(value) {
    return typeof value === "string";
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
class DynamicFormControlComponent {
    /**
     * @protected
     * @param {?} layoutService
     * @param {?} validationService
     */
    constructor(layoutService, validationService) {
        this.layoutService = layoutService;
        this.validationService = validationService;
        this._hasFocus = false;
    }
    /**
     * @return {?}
     */
    get control() {
        /** @type {?} */
        const control = this.group.get(this.model.id);
        if (control === null) {
            throw new Error(`form group does not contain an abstract control with id ${this.model.id}`);
        }
        return (/** @type {?} */ (control));
    }
    /**
     * @return {?}
     */
    get elementId() {
        return this.layoutService.getElementId(this.model);
    }
    /**
     * @return {?}
     */
    get errorMessages() {
        return this.validationService.createErrorMessages(this.control, this.model);
    }
    /**
     * @return {?}
     */
    get hasFocus() {
        return this._hasFocus;
    }
    /**
     * @return {?}
     */
    get isInvalid() {
        return this.control.invalid;
    }
    /**
     * @return {?}
     */
    get isValid() {
        return this.control.valid;
    }
    /**
     * @return {?}
     */
    get showErrorMessages() {
        return this.model.hasErrorMessages && this.control.touched && !this.hasFocus && this.isInvalid;
    }
    /**
     * @param {?} context
     * @param {?} place
     * @param {?=} model
     * @return {?}
     */
    getClass(context, place, model = this.model) {
        /** @type {?} */
        let controlLayout = this.layoutService.findByModel(model, this.layout) || (/** @type {?} */ (model.layout));
        return this.layoutService.getClass(controlLayout, context, place);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onBlur($event) {
        if ($event instanceof Event) {
            $event.stopPropagation();
        }
        this._hasFocus = false;
        this.blur.emit($event);
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onChange($event) {
        if ($event instanceof Event) {
            $event.stopPropagation();
        }
        this.change.emit($event);
    }
    /**
     * @param {?} $event
     * @param {?} model
     * @return {?}
     */
    onEmbeddedCheckboxChange($event, model) {
        this.onChange($event);
        model.value = ((/** @type {?} */ ($event.target))).checked;
    }
    /**
     * @param {?} $event
     * @param {?=} type
     * @param {?=} bypass
     * @return {?}
     */
    onCustomEvent($event, type = null, bypass = false) {
        /** @type {?} */
        let emitter = (/** @type {?} */ (this.customEvent));
        if (bypass) {
            emitter.emit($event);
        }
        else if (isString(type)) {
            emitter.emit({ customEvent: $event, customEventType: type });
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onFocus($event) {
        if ($event instanceof Event) {
            $event.stopPropagation();
        }
        this._hasFocus = true;
        this.focus.emit($event);
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
class DynamicFormArrayComponent extends DynamicFormControlComponent {
    /**
     * @return {?}
     */
    get array() {
        return (/** @type {?} */ (this.control));
    }
    /**
     * @return {?}
     */
    get startTemplate() {
        return this.layoutService.getStartTemplate(this.model, this.templates);
    }
    /**
     * @return {?}
     */
    get endTemplate() {
        return this.layoutService.getEndTemplate(this.model, this.templates);
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const DynamicFormControlEventType = {
    Blur: "blur", Change: "change", Custom: "custom", Focus: "focus",
};
/**
 * @param {?} $event
 * @return {?}
 */
function isDynamicFormControlEvent($event) {
    return isObject($event) && $event.hasOwnProperty("$event");
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
class DynamicFormComponent {
    /**
     * @protected
     * @param {?} formService
     * @param {?} layoutService
     */
    constructor(formService, layoutService) {
        this.formService = formService;
        this.layoutService = layoutService;
    }
    /**
     * @param {?} _index
     * @param {?} model
     * @return {?}
     */
    trackByFn(_index, model) {
        return model.id;
    }
    /**
     * @param {?} model
     * @param {?} context
     * @param {?} place
     * @return {?}
     */
    getClass(model, context, place) {
        /** @type {?} */
        let controlLayout = this.layoutService.findByModel(model, this.formLayout) || (/** @type {?} */ (model.layout));
        return this.layoutService.getClass(controlLayout, context, place);
    }
    /**
     * @param {?} $event
     * @param {?} type
     * @return {?}
     */
    onEvent($event, type) {
        switch (type) {
            case DynamicFormControlEventType.Blur:
                this.blur.emit($event);
                break;
            case DynamicFormControlEventType.Change:
                this.change.emit($event);
                break;
            case DynamicFormControlEventType.Focus:
                this.focus.emit($event);
                break;
            case DynamicFormControlEventType.Custom:
                this.customEvent.emit($event);
                break;
        }
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const METADATA_KEY_SERIALIZABLE = "SERIALIZABLE";
/**
 * @param {?=} name
 * @return {?}
 */
function serializable(name) {
    return (/**
     * @param {?} target
     * @param {?} key
     * @return {?}
     */
    function (target, key) {
        Reflect.defineMetadata(METADATA_KEY_SERIALIZABLE, { key: key, name: name || key }, target, key);
    });
}
/**
 * @param {?} target
 * @return {?}
 */
function getSerializables(target) {
    /** @type {?} */
    let serializables = [];
    for (let key in target) {
        /** @type {?} */
        let metadata = Reflect.getMetadata(METADATA_KEY_SERIALIZABLE, target, key);
        if (metadata) {
            serializables.push(metadata);
        }
    }
    return serializables;
}
/**
 * @param {?} target
 * @param {?=} prototype
 * @return {?}
 */
function serialize(target, prototype) {
    return getSerializables(prototype || target).reduce((/**
     * @param {?} prev
     * @param {?} prop
     * @return {?}
     */
    (prev, prop) => {
        prev[prop.name] = target[prop.key];
        return prev;
    }), {});
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
class DynamicFormControlModel {
    /**
     * @protected
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout = null) {
        this.parent = null;
        this.asyncValidators = config.asyncValidators || null;
        this.errorMessages = config.errorMessages || null;
        this.hidden = isBoolean(config.hidden) ? config.hidden : false;
        this.id = config.id;
        this.label = config.label || null;
        this.labelTooltip = config.labelTooltip || null;
        this.controlTooltip = config.controlTooltip || null;
        this.layout = layout;
        this.name = config.name || config.id;
        this.relations = Array.isArray(config.relations) ? config.relations : [];
        this.updateOn = isString(config.updateOn) ? config.updateOn : null;
        this.validators = config.validators || null;
        this.disabled$ = new BehaviorSubject(isBoolean(config.disabled) ? config.disabled : false);
        this.disabled$.subscribe((/**
         * @param {?} disabled
         * @return {?}
         */
        disabled => this._disabled = disabled));
        this.disabledChanges = this.disabled$.asObservable();
    }
    /**
     * @return {?}
     */
    get disabled() {
        return this.disabled$.getValue();
    }
    /**
     * @param {?} disabled
     * @return {?}
     */
    set disabled(disabled) {
        this.disabled$.next(disabled);
    }
    /**
     * @return {?}
     */
    get hasErrorMessages() {
        return isObject(this.errorMessages);
    }
    /**
     * @return {?}
     */
    toJSON() {
        return serialize(this);
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicFormControlModel.prototype, "asyncValidators", void 0);
__decorate([
    serializable("disabled"),
    __metadata("design:type", Boolean)
], DynamicFormControlModel.prototype, "_disabled", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicFormControlModel.prototype, "errorMessages", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicFormControlModel.prototype, "hidden", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFormControlModel.prototype, "id", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFormControlModel.prototype, "label", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFormControlModel.prototype, "labelTooltip", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFormControlModel.prototype, "controlTooltip", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicFormControlModel.prototype, "layout", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFormControlModel.prototype, "name", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Array)
], DynamicFormControlModel.prototype, "relations", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFormControlModel.prototype, "updateOn", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicFormControlModel.prototype, "validators", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 * @template T
 */
class DynamicFormValueControlModel extends DynamicFormControlModel {
    /**
     * @protected
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.additional = isObject(config.additional) ? config.additional : null;
        this.hint = config.hint || null;
        this.required = isBoolean(config.required) ? config.required : false;
        this.tabIndex = config.tabIndex || null;
        this.value$ = new BehaviorSubject(config.value !== null && config.value !== undefined ? config.value : null);
        this.value$.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => this._value = value));
        this.valueChanges = this.value$.asObservable();
    }
    /**
     * @return {?}
     */
    get value() {
        return this.value$.getValue();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set value(value) {
        this.value$.next(value);
    }
    /**
     * @param {?} key
     * @param {?=} defaultValue
     * @return {?}
     */
    getAdditional(key, defaultValue = undefined) {
        return this.additional !== null && this.additional.hasOwnProperty(key) ? this.additional[key] : defaultValue;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicFormValueControlModel.prototype, "additional", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFormValueControlModel.prototype, "hint", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicFormValueControlModel.prototype, "required", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicFormValueControlModel.prototype, "tabIndex", void 0);
__decorate([
    serializable("value"),
    __metadata("design:type", Object)
], DynamicFormValueControlModel.prototype, "_value", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DynamicFormArrayGroupModel {
    /**
     * @param {?} context
     * @param {?=} group
     * @param {?=} index
     */
    constructor(context, group = [], index = -1) {
        this.$implicit = this;
        this.context = context;
        this.group = group;
        this.index = index;
    }
    /**
     * @return {?}
     */
    get parent() {
        return this.context;
    }
    /**
     * @param {?} index
     * @return {?}
     */
    get(index) {
        return this.group[index];
    }
    /**
     * @return {?}
     */
    toJSON() {
        return serialize(this);
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Array)
], DynamicFormArrayGroupModel.prototype, "group", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicFormArrayGroupModel.prototype, "index", void 0);
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_ARRAY = "ARRAY";
class DynamicFormArrayModel extends DynamicFormControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.groups = [];
        // only to recreate model from JSON
        this.type = DYNAMIC_FORM_CONTROL_TYPE_ARRAY;
        if (isFunction(config.groupFactory)) {
            this.groupFactory = config.groupFactory;
        }
        else {
            throw new Error("group factory function must be specified for DynamicFormArrayModel");
        }
        this.groupAsyncValidators = config.groupAsyncValidators || null;
        this.groupPrototype = this.groupFactory();
        this.groupValidators = config.groupValidators || null;
        this.initialCount = isNumber(config.initialCount) ? config.initialCount : 1;
        if (Array.isArray(config.groups)) {
            config.groups.forEach((/**
             * @param {?} arrayGroup
             * @param {?} index
             * @return {?}
             */
            (arrayGroup, index) => {
                this.groups.push(new DynamicFormArrayGroupModel(this, arrayGroup.group, arrayGroup.index || index));
            }));
        }
        else {
            for (let index = 0; index < this.initialCount; index++) {
                this.addGroup();
            }
        }
    }
    /**
     * @private
     * @return {?}
     */
    updateGroupIndex() {
        this.groups.forEach((/**
         * @param {?} group
         * @param {?} index
         * @return {?}
         */
        (group, index) => group.index = index));
    }
    /**
     * @return {?}
     */
    get size() {
        return this.groups.length;
    }
    /**
     * @param {?} index
     * @return {?}
     */
    get(index) {
        return this.groups[index];
    }
    /**
     * @return {?}
     */
    addGroup() {
        return this.insertGroup(this.groups.length);
    }
    /**
     * @param {?} index
     * @return {?}
     */
    insertGroup(index) {
        /** @type {?} */
        let group = new DynamicFormArrayGroupModel(this, this.groupFactory());
        this.groups.splice(index, 0, group);
        this.updateGroupIndex();
        return group;
    }
    /**
     * @param {?} index
     * @param {?} step
     * @return {?}
     */
    moveGroup(index, step) {
        this.groups.splice(index + step, 0, ...this.groups.splice(index, 1));
        this.updateGroupIndex();
    }
    /**
     * @param {?} index
     * @return {?}
     */
    removeGroup(index) {
        this.groups.splice(index, 1);
        this.updateGroupIndex();
    }
    /**
     * @return {?}
     */
    clear() {
        this.groups.splice(0);
        this.updateGroupIndex();
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicFormArrayModel.prototype, "groupAsyncValidators", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicFormArrayModel.prototype, "groupValidators", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Array)
], DynamicFormArrayModel.prototype, "groups", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicFormArrayModel.prototype, "initialCount", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Array)
], DynamicFormArrayModel.prototype, "groupPrototype", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFormArrayModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
class DynamicCheckControlModel extends DynamicFormValueControlModel {
    /**
     * @protected
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.labelPosition = config.labelPosition || null;
        this.checked = isBoolean(this.value) ? this.value : false;
    }
    /**
     * @return {?}
     */
    get checked() {
        return this.value;
    }
    /**
     * @param {?} checked
     * @return {?}
     */
    set checked(checked) {
        this.value = checked;
    }
    /**
     * @return {?}
     */
    toggle() {
        this.checked = !this.checked;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicCheckControlModel.prototype, "labelPosition", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX = "CHECKBOX";
class DynamicCheckboxModel extends DynamicCheckControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX;
        this.indeterminate = isBoolean(config.indeterminate) ? config.indeterminate : false;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicCheckboxModel.prototype, "indeterminate", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicCheckboxModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const AUTOCOMPLETE_OFF = "off";
/** @type {?} */
const AUTOCOMPLETE_ON = "on";
/** @type {?} */
const AUTOFILL_TOKEN_BILLING = "billing";
/** @type {?} */
const AUTOFILL_TOKEN_SHIPPING = "shipping";
/** @type {?} */
const AUTOFILL_TOKENS_ADDRESS = [AUTOFILL_TOKEN_BILLING, AUTOFILL_TOKEN_SHIPPING];
/** @type {?} */
const AUTOFILL_TOKEN_HOME = "home";
/** @type {?} */
const AUTOFILL_TOKEN_WORK = "work";
/** @type {?} */
const AUTOFILL_TOKEN_MOBILE = "mobile";
/** @type {?} */
const AUTOFILL_TOKEN_FAX = "fax";
/** @type {?} */
const AUTOFILL_TOKEN_PAGER = "pager";
/** @type {?} */
const AUTOFILL_TOKENS_CONTACT = [
    AUTOFILL_TOKEN_FAX, AUTOFILL_TOKEN_HOME, AUTOFILL_TOKEN_MOBILE, AUTOFILL_TOKEN_PAGER, AUTOFILL_TOKEN_WORK
];
/** @type {?} */
const AUTOFILL_FIELD_STREET_ADDRESS = "street-address";
/** @type {?} */
const AUTOFILL_FIELD_ADDRESS_LINE_1 = "address-line1";
/** @type {?} */
const AUTOFILL_FIELD_ADDRESS_LINE_2 = "address-line2";
/** @type {?} */
const AUTOFILL_FIELD_ADDRESS_LINE_3 = "address-line3";
/** @type {?} */
const AUTOFILL_FIELD_ADDRESS_LEVEL_4 = "address-level4";
/** @type {?} */
const AUTOFILL_FIELD_ADDRESS_LEVEL_3 = "address-level3";
/** @type {?} */
const AUTOFILL_FIELD_ADDRESS_LEVEL_2 = "address-level2";
/** @type {?} */
const AUTOFILL_FIELD_ADDRESS_LEVEL_1 = "address-level1";
/** @type {?} */
const AUTOFILL_FIELD_NAME = "name";
/** @type {?} */
const AUTOFILL_FIELD_HONORIFIC_PREFIX = "honorific-prefix";
/** @type {?} */
const AUTOFILL_FIELD_GIVEN_NAME = "given-name";
/** @type {?} */
const AUTOFILL_FIELD_ADDITIONAL_NAME = "additional-name";
/** @type {?} */
const AUTOFILL_FIELD_FAMILY_NAME = "family-name";
/** @type {?} */
const AUTOFILL_FIELD_HONORIFIC_SUFFIX = "honorific-suffix";
/** @type {?} */
const AUTOFILL_FIELD_NICKNAME = "nickname";
/** @type {?} */
const AUTOFILL_FIELD_USERNAME = "username";
/** @type {?} */
const AUTOFILL_FIELD_NEW_PASSWORD = "new-password";
/** @type {?} */
const AUTOFILL_FIELD_CURRENT_PASSWORD = "current-password";
/** @type {?} */
const AUTOFILL_FIELD_ORGANIZATION_TITLE = "organization-title";
/** @type {?} */
const AUTOFILL_FIELD_ORGANIZATION = "organization";
/** @type {?} */
const AUTOFILL_FIELD_COUNTRY = "country";
/** @type {?} */
const AUTOFILL_FIELD_COUNTRY_NAME = "country-name";
/** @type {?} */
const AUTOFILL_FIELD_POSTAL_CODE = "postal-code";
/** @type {?} */
const AUTOFILL_FIELD_CC_NAME = "cc-name";
/** @type {?} */
const AUTOFILL_FIELD_CC_GIVEN_NAME = "cc-given-name";
/** @type {?} */
const AUTOFILL_FIELD_CC_ADDITIONAL_NAME = "cc-additional-name";
/** @type {?} */
const AUTOFILL_FIELD_CC_FAMILY_NAME = "cc-family-name";
/** @type {?} */
const AUTOFILL_FIELD_CC_NUMBER = "cc-number";
/** @type {?} */
const AUTOFILL_FIELD_CC_EXP = "cc-exp";
/** @type {?} */
const AUTOFILL_FIELD_CC_EXP_MONTH = "cc-exp-month";
/** @type {?} */
const AUTOFILL_FIELD_CC_EXP_YEAR = "cc-exp-year";
/** @type {?} */
const AUTOFILL_FIELD_CC_CSC = "cc-csc";
/** @type {?} */
const AUTOFILL_FIELD_CC_TYPE = "cc-type";
/** @type {?} */
const AUTOFILL_FIELD_TRANSACTION_CURRENCY = "transaction-currency";
/** @type {?} */
const AUTOFILL_FIELD_TRANSACTION_AMOUNT = "transaction-amount";
/** @type {?} */
const AUTOFILL_FIELD_LANGUAGE = "language";
/** @type {?} */
const AUTOFILL_FIELD_BDAY = "bday";
/** @type {?} */
const AUTOFILL_FIELD_BDAY_DAY = "bday-day";
/** @type {?} */
const AUTOFILL_FIELD_BDAY_MONTH = "bday-month";
/** @type {?} */
const AUTOFILL_FIELD_BDAY_YEAR = "bday-year";
/** @type {?} */
const AUTOFILL_FIELD_SEX = "sex";
/** @type {?} */
const AUTOFILL_FIELD_URL = "url";
/** @type {?} */
const AUTOFILL_FIELD_PHOTO = "photo";
/** @type {?} */
const AUTOFILL_FIELDS = [
    AUTOFILL_FIELD_STREET_ADDRESS, AUTOFILL_FIELD_ADDRESS_LINE_1, AUTOFILL_FIELD_ADDRESS_LINE_2,
    AUTOFILL_FIELD_ADDRESS_LINE_3, AUTOFILL_FIELD_ADDRESS_LEVEL_4, AUTOFILL_FIELD_ADDRESS_LEVEL_3,
    AUTOFILL_FIELD_ADDRESS_LEVEL_2, AUTOFILL_FIELD_ADDRESS_LEVEL_1, AUTOFILL_FIELD_NAME,
    AUTOFILL_FIELD_HONORIFIC_PREFIX, AUTOFILL_FIELD_GIVEN_NAME, AUTOFILL_FIELD_ADDITIONAL_NAME,
    AUTOFILL_FIELD_FAMILY_NAME, AUTOFILL_FIELD_HONORIFIC_SUFFIX, AUTOFILL_FIELD_NICKNAME, AUTOFILL_FIELD_USERNAME,
    AUTOFILL_FIELD_NEW_PASSWORD, AUTOFILL_FIELD_CURRENT_PASSWORD, AUTOFILL_FIELD_ORGANIZATION_TITLE,
    AUTOFILL_FIELD_ORGANIZATION, AUTOFILL_FIELD_COUNTRY, AUTOFILL_FIELD_COUNTRY_NAME, AUTOFILL_FIELD_POSTAL_CODE,
    AUTOFILL_FIELD_CC_NAME, AUTOFILL_FIELD_CC_GIVEN_NAME, AUTOFILL_FIELD_CC_ADDITIONAL_NAME,
    AUTOFILL_FIELD_CC_FAMILY_NAME, AUTOFILL_FIELD_CC_NUMBER, AUTOFILL_FIELD_CC_EXP, AUTOFILL_FIELD_CC_EXP_MONTH,
    AUTOFILL_FIELD_CC_EXP_YEAR, AUTOFILL_FIELD_CC_CSC, AUTOFILL_FIELD_CC_TYPE, AUTOFILL_FIELD_TRANSACTION_CURRENCY,
    AUTOFILL_FIELD_TRANSACTION_AMOUNT, AUTOFILL_FIELD_LANGUAGE, AUTOFILL_FIELD_BDAY, AUTOFILL_FIELD_BDAY_DAY,
    AUTOFILL_FIELD_BDAY_MONTH, AUTOFILL_FIELD_BDAY_YEAR, AUTOFILL_FIELD_SEX, AUTOFILL_FIELD_URL, AUTOFILL_FIELD_PHOTO
];
/** @type {?} */
const AUTOFILL_FIELD_TEL = "tel";
/** @type {?} */
const AUTOFILL_FIELD_TEL_COUNTRY_CODE = "tel-country-code";
/** @type {?} */
const AUTOFILL_FIELD_TEL_NATIONAL = "tel-national";
/** @type {?} */
const AUTOFILL_FIELD_TEL_AREA_CODE = "tel-area-code";
/** @type {?} */
const AUTOFILL_FIELD_TEL_LOCAL = "tel-local";
/** @type {?} */
const AUTOFILL_FIELD_TEL_LOCAL_PREFIX = "tel-local-prefix";
/** @type {?} */
const AUTOFILL_FIELD_TEL_LOCAL_SUFFIX = "tel-local-suffix";
/** @type {?} */
const AUTOFILL_FIELD_TEL_LOCAL_EXTENSION = "tel-extension";
/** @type {?} */
const AUTOFILL_FIELD_EMAIL = "email";
/** @type {?} */
const AUTOFILL_FIELD_IMPP = "impp";
/** @type {?} */
const AUTOFILL_FIELDS_CONTACT = [
    AUTOFILL_FIELD_TEL, AUTOFILL_FIELD_TEL_COUNTRY_CODE, AUTOFILL_FIELD_TEL_NATIONAL, AUTOFILL_FIELD_TEL_AREA_CODE,
    AUTOFILL_FIELD_TEL_LOCAL, AUTOFILL_FIELD_TEL_LOCAL_PREFIX, AUTOFILL_FIELD_TEL_LOCAL_SUFFIX,
    AUTOFILL_FIELD_TEL_LOCAL_EXTENSION, AUTOFILL_FIELD_EMAIL, AUTOFILL_FIELD_IMPP
];
/**
 * @param {?} token
 * @return {?}
 */
function isAddressToken(token) {
    return AUTOFILL_TOKENS_ADDRESS.indexOf(token) > -1;
}
/**
 * @param {?} token
 * @return {?}
 */
function isContactField(token) {
    return AUTOFILL_FIELDS_CONTACT.indexOf(token) > -1;
}
/**
 * @param {?} token
 * @return {?}
 */
function isContactToken(token) {
    return AUTOFILL_TOKENS_CONTACT.indexOf(token) > -1;
}
/**
 * @param {?} token
 * @return {?}
 */
function isField(token) {
    return AUTOFILL_FIELDS.indexOf(token) > -1;
}
/**
 * @param {?} token
 * @return {?}
 */
function isSectionToken(token) {
    return token.startsWith("section-");
}
/**
 * @param {?} tokens
 * @return {?}
 */
function validate(tokens) {
    /** @type {?} */
    let toExpression = (/**
     * @param {?} total
     * @param {?} currentValue
     * @return {?}
     */
    (total, currentValue) => `${total}|${currentValue}`);
    /** @type {?} */
    let tokensAddress = AUTOFILL_TOKENS_ADDRESS.reduce(toExpression);
    /** @type {?} */
    let tokensContact = AUTOFILL_TOKENS_CONTACT.reduce(toExpression);
    /** @type {?} */
    let fields = AUTOFILL_FIELDS.reduce(toExpression);
    /** @type {?} */
    let fieldsContact = AUTOFILL_FIELDS_CONTACT.reduce(toExpression);
    /** @type {?} */
    let regex = new RegExp(`^(section-\\w+\\s{1})?((${tokensAddress}){1}\\s)?((${fields}){1}|((${tokensContact}){1}\\s{1}(${fieldsContact})))$`);
    return regex.test(tokens);
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 * @template T
 */
class DynamicInputControlModel extends DynamicFormValueControlModel {
    /**
     * @protected
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.autoComplete = config.autoComplete || AUTOCOMPLETE_ON;
        this.autoFocus = isBoolean(config.autoFocus) ? config.autoFocus : false;
        this.maxLength = isNumber(config.maxLength) ? config.maxLength : null;
        this.minLength = isNumber(config.minLength) ? config.minLength : null;
        this.placeholder = config.placeholder || "";
        this.prefix = config.prefix || null;
        this.readOnly = isBoolean(config.readOnly) ? config.readOnly : false;
        this.spellCheck = isBoolean(config.spellCheck) ? config.spellCheck : false;
        this.suffix = config.suffix || null;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicInputControlModel.prototype, "autoComplete", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicInputControlModel.prototype, "autoFocus", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicInputControlModel.prototype, "maxLength", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicInputControlModel.prototype, "minLength", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicInputControlModel.prototype, "placeholder", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicInputControlModel.prototype, "prefix", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicInputControlModel.prototype, "readOnly", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicInputControlModel.prototype, "spellCheck", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicInputControlModel.prototype, "suffix", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @param {?} mask
 * @return {?}
 */
function maskToString(mask) {
    if (isString(mask)) {
        return (/** @type {?} */ (mask));
    }
    else if (mask instanceof RegExp) {
        return mask.toString();
    }
    else if (Array.isArray(mask)) {
        return (/** @type {?} */ (mask.map((/**
         * @param {?} value
         * @return {?}
         */
        value => maskToString(value)))));
    }
    return null;
}
/**
 * @param {?} mask
 * @return {?}
 */
function maskFromString(mask) {
    if (isString(mask)) {
        /** @type {?} */
        let isRegExp = ((/** @type {?} */ (mask))).startsWith("/") && ((/** @type {?} */ (mask))).endsWith("/");
        return isRegExp ? new RegExp(((/** @type {?} */ (mask))).slice(1, mask.length - 1)) : mask;
    }
    else if (Array.isArray(mask)) {
        return (/** @type {?} */ (((/** @type {?} */ (mask))).map((/**
         * @param {?} value
         * @return {?}
         */
        value => maskFromString(value)))));
    }
    return null;
}
/**
 * @param {?} _key
 * @param {?} value
 * @return {?}
 */
function parseReviver(_key, value) {
    /** @type {?} */
    let regexDateISO = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2}(?:\.\d*))(?:Z|([+\-])([\d|:]*))?$/;
    return isString(value) && regexDateISO.test(value) ? new Date(value) : value;
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_INPUT = "INPUT";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_COLOR = "color";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_DATE = "date";
//export const DYNAMIC_FORM_CONTROL_INPUT_TYPE_DATETIME = "datetime";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_DATETIME_LOCAL = "datetime-local";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_EMAIL = "email";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_FILE = "file";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_MONTH = "month";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_NUMBER = "number";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_PASSWORD = "password";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_RANGE = "range";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_SEARCH = "search";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_TEL = "tel";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_TEXT = "text";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_TIME = "time";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_URL = "url";
/** @type {?} */
const DYNAMIC_FORM_CONTROL_INPUT_TYPE_WEEK = "week";
class DynamicInputModel extends DynamicInputControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.files = null;
        this.list$ = null;
        this._list = null;
        this._listId = null;
        this.type = DYNAMIC_FORM_CONTROL_TYPE_INPUT;
        this.accept = config.accept || null;
        this.inputType = config.inputType || DYNAMIC_FORM_CONTROL_INPUT_TYPE_TEXT;
        this.mask = config.mask || null;
        this.max = config.max !== undefined ? config.max : null;
        this.min = config.min !== undefined ? config.min : null;
        this.multiple = isBoolean(config.multiple) ? config.multiple : null;
        this.pattern = config.pattern || null;
        this.step = isNumber(config.step) ? config.step : null;
        if (config.list !== undefined) {
            this.list = config.list;
            this._listId = `${this.id}List`;
        }
    }
    /**
     * @return {?}
     */
    get listId() {
        return this._listId;
    }
    /**
     * @return {?}
     */
    get hasList() {
        return isObservable(this.list$);
    }
    /**
     * @param {?} list
     * @return {?}
     */
    set list(list) {
        if (Array.isArray(list)) {
            this._list = list;
            this.list$ = of(this._list);
        }
        else if (isObservable(list)) {
            this.list$ = ((/** @type {?} */ (list))).pipe(tap((/**
             * @param {?} list
             * @return {?}
             */
            list => this._list = list)));
        }
        else {
            this._list = null;
            this.list$ = null;
        }
    }
    /**
     * @return {?}
     */
    toJSON() {
        /** @type {?} */
        let json = super.toJSON();
        if (this.mask !== null) {
            if (this.mask instanceof Function) {
                json.mask = this.mask;
            }
            else {
                json.mask = maskToString(this.mask);
            }
        }
        return json;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicInputModel.prototype, "accept", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicInputModel.prototype, "inputType", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicInputModel.prototype, "mask", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicInputModel.prototype, "max", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicInputModel.prototype, "min", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicInputModel.prototype, "multiple", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicInputModel.prototype, "pattern", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicInputModel.prototype, "step", void 0);
__decorate([
    serializable("list"),
    __metadata("design:type", Array)
], DynamicInputModel.prototype, "_list", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicInputModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
class DynamicFormControlContainerComponent {
    /**
     * @protected
     * @param {?} componentFactoryResolver
     * @param {?} layoutService
     * @param {?} validationService
     * @param {?} componentService
     * @param {?} relationService
     */
    constructor(componentFactoryResolver, layoutService, validationService, componentService, relationService) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.layoutService = layoutService;
        this.validationService = validationService;
        this.componentService = componentService;
        this.relationService = relationService;
        this.context = null;
        this.componentSubscriptions = [];
        this.subscriptions = [];
    }
    /**
     * @param {?} changes
     * @return {?}
     */
    ngOnChanges(changes) {
        /** @type {?} */
        const groupChange = (/** @type {?} */ (changes["group"]));
        /** @type {?} */
        const modelChange = (/** @type {?} */ (changes["model"]));
        if (modelChange) {
            this.destroyFormControlComponent();
            this.createFormControlComponent();
        }
        if (groupChange || modelChange) {
            if (this.model) {
                this.unsubscribe();
                if (this.group) {
                    this.control = (/** @type {?} */ (this.group.get(this.model.id)));
                    this.subscriptions.push(this.control.valueChanges.subscribe((/**
                     * @param {?} value
                     * @return {?}
                     */
                    value => this.onControlValueChanges(value))));
                }
                this.subscriptions.push(this.model.disabledChanges.subscribe((/**
                 * @param {?} value
                 * @return {?}
                 */
                value => this.onModelDisabledUpdates(value))));
                if (this.model instanceof DynamicFormValueControlModel) {
                    /** @type {?} */
                    const model = (/** @type {?} */ (this.model));
                    this.subscriptions.push(model.valueChanges.subscribe((/**
                     * @param {?} value
                     * @return {?}
                     */
                    value => this.onModelValueUpdates(value))));
                }
                if (this.model.relations.length > 0) {
                    this.subscriptions.push(...this.relationService.subscribeRelations(this.model, this.group, this.control));
                }
            }
        }
    }
    /**
     * @return {?}
     */
    ngOnDestroy() {
        this.destroyFormControlComponent();
        this.unsubscribe();
    }
    /**
     * @return {?}
     */
    get errorMessages() {
        return this.validationService.createErrorMessages(this.control, this.model);
    }
    /**
     * @return {?}
     */
    get hasHint() {
        return isString(((/** @type {?} */ (this.model))).hint);
    }
    /**
     * @return {?}
     */
    get hint() {
        return ((/** @type {?} */ (this.model))).hint || null;
    }
    /**
     * @return {?}
     */
    get hasLabel() {
        return isString(this.model.label);
    }
    /**
     * @return {?}
     */
    get isCheckbox() {
        return this.model.type === DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX;
    }
    /**
     * @return {?}
     */
    get elementId() {
        return this.layoutService.getElementId(this.model);
    }
    /**
     * @return {?}
     */
    get isInvalid() {
        return this.control.invalid;
    }
    /**
     * @return {?}
     */
    get isValid() {
        return this.control.valid;
    }
    /**
     * @return {?}
     */
    get showErrorMessages() {
        return this.model.hasErrorMessages && this.control.touched && !this.hasFocus && this.isInvalid;
    }
    /**
     * @return {?}
     */
    get templates() {
        return this.inputTemplateList !== undefined ? this.inputTemplateList : this.contentTemplateList;
    }
    /**
     * @return {?}
     */
    get startTemplate() {
        return this.model.type !== DYNAMIC_FORM_CONTROL_TYPE_ARRAY ?
            this.layoutService.getStartTemplate(this.model, this.templates) : undefined;
    }
    /**
     * @return {?}
     */
    get endTemplate() {
        return this.model.type !== DYNAMIC_FORM_CONTROL_TYPE_ARRAY ?
            this.layoutService.getEndTemplate(this.model, this.templates) : undefined;
    }
    /**
     * @param {?} context
     * @param {?} place
     * @param {?=} model
     * @return {?}
     */
    getClass(context, place, model = this.model) {
        /** @type {?} */
        let controlLayout = this.layoutService.findByModel(model, this.layout) || (/** @type {?} */ (model.layout));
        return this.layoutService.getClass(controlLayout, context, place);
    }
    /**
     * @protected
     * @return {?}
     */
    createFormControlComponent() {
        /** @type {?} */
        let componentType = this.componentType;
        if (componentType !== null) {
            /** @type {?} */
            let componentFactory = this.componentFactoryResolver.resolveComponentFactory(componentType);
            this.componentViewContainerRef.clear();
            this.componentRef = this.componentViewContainerRef.createComponent(componentFactory);
            /** @type {?} */
            let instance = this.componentRef.instance;
            instance.group = this.group;
            instance.layout = this.layout;
            instance.model = (/** @type {?} */ (this.model));
            if (this.templates) {
                instance.templates = this.templates;
            }
            this.componentSubscriptions.push(instance.blur.subscribe((/**
             * @param {?} $event
             * @return {?}
             */
            ($event) => this.onBlur($event))));
            this.componentSubscriptions.push(instance.change.subscribe((/**
             * @param {?} $event
             * @return {?}
             */
            ($event) => this.onChange($event))));
            this.componentSubscriptions.push(instance.focus.subscribe((/**
             * @param {?} $event
             * @return {?}
             */
            ($event) => this.onFocus($event))));
            if (instance.customEvent !== undefined) {
                this.componentSubscriptions.push(instance.customEvent.subscribe((/**
                 * @param {?} $event
                 * @return {?}
                 */
                ($event) => this.onCustomEvent($event))));
            }
            this.registerComponentRef(this.componentRef);
        }
    }
    /**
     * @protected
     * @return {?}
     */
    destroyFormControlComponent() {
        if (this.componentRef) {
            this.componentSubscriptions.forEach((/**
             * @param {?} subscription
             * @return {?}
             */
            subscription => subscription.unsubscribe()));
            this.componentSubscriptions = [];
            this.unregisterComponentRef();
            this.componentRef.destroy();
        }
    }
    /**
     * @protected
     * @param {?} $event
     * @param {?} type
     * @return {?}
     */
    createDynamicFormControlEvent($event, type) {
        return { $event, context: this.context, control: this.control, group: this.group, model: this.model, type };
    }
    /**
     * @return {?}
     */
    unsubscribe() {
        this.subscriptions.forEach((/**
         * @param {?} subscription
         * @return {?}
         */
        subscription => subscription.unsubscribe()));
        this.subscriptions = [];
    }
    /**
     * @param {?} value
     * @return {?}
     */
    onControlValueChanges(value) {
        if (this.model instanceof DynamicFormValueControlModel && this.model.value !== value) {
            this.model.value = value;
        }
    }
    /**
     * @param {?} value
     * @return {?}
     */
    onModelValueUpdates(value) {
        if (this.control.value !== value) {
            this.control.setValue(value);
        }
    }
    /**
     * @param {?} disabled
     * @return {?}
     */
    onModelDisabledUpdates(disabled) {
        disabled ? this.control.disable() : this.control.enable();
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onChange($event) {
        if ($event instanceof Event) { // native HTML5 change event
            if (this.model.type === DYNAMIC_FORM_CONTROL_TYPE_INPUT) {
                /** @type {?} */
                let model = (/** @type {?} */ (this.model));
                if (model.inputType === DYNAMIC_FORM_CONTROL_INPUT_TYPE_FILE) {
                    /** @type {?} */
                    let inputElement = $event.target || $event.srcElement;
                    model.files = (/** @type {?} */ (inputElement.files));
                }
            }
            this.change.emit(this.createDynamicFormControlEvent($event, DynamicFormControlEventType.Change));
        }
        else if (isDynamicFormControlEvent($event)) { // event bypass
            this.change.emit($event);
        }
        else { // custom library value change event
            this.change.emit(this.createDynamicFormControlEvent($event, DynamicFormControlEventType.Change));
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onBlur($event) {
        if (isDynamicFormControlEvent($event)) { // event bypass
            this.blur.emit($event);
        }
        else { // native HTML 5 or UI library blur event
            this.hasFocus = false;
            this.blur.emit(this.createDynamicFormControlEvent($event, DynamicFormControlEventType.Blur));
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onFocus($event) {
        if (isDynamicFormControlEvent($event)) { // event bypass
            this.focus.emit($event);
        }
        else { // native HTML 5 or UI library focus event
            this.hasFocus = true;
            this.focus.emit(this.createDynamicFormControlEvent($event, DynamicFormControlEventType.Focus));
        }
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    onCustomEvent($event) {
        /** @type {?} */
        let emitter = (/** @type {?} */ (this.customEvent));
        if (isDynamicFormControlEvent($event)) { // child event bypass
            emitter.emit($event);
        }
        else { // native UI library custom event
            emitter.emit(this.createDynamicFormControlEvent($event.customEvent, $event.customEventType));
        }
    }
    /**
     * @private
     * @param {?} instanceRef
     * @return {?}
     */
    registerComponentRef(instanceRef) {
        /** @type {?} */
        let index;
        if (this.context instanceof DynamicFormArrayGroupModel) {
            index = this.context.index;
        }
        this.componentService.registerFormControlRef(this.model, instanceRef, index);
    }
    /**
     * @private
     * @return {?}
     */
    unregisterComponentRef() {
        /** @type {?} */
        let index;
        if (this.context instanceof DynamicFormArrayGroupModel) {
            index = this.context.index;
        }
        this.componentService.unregisterFormControlRef(this.model.id, index);
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
class DynamicTemplateableFormControlComponent extends DynamicFormControlComponent {
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        this.layoutService
            .filterTemplatesByModel(this.model, this.templates)
            .forEach((/**
         * @param {?} template
         * @return {?}
         */
        template => this.bindTemplate(template)));
    }
    /**
     * @param {?} template
     * @return {?}
     */
    bindTemplate(template) {
        if (isString(template.as) && this.templateDirectives.has(template.as)) {
            /** @type {?} */
            let property = (/** @type {?} */ (this.templateDirectives.get(template.as)));
            this.viewChild[property] = this.mapTemplate(template);
        }
    }
}

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DynamicIdDirective {
    /**
     * @param {?} elementRef
     * @param {?} renderer
     */
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (this.dynamicId) {
            this.renderer.setAttribute(this.elementRef.nativeElement, "id", (/** @type {?} */ (this.dynamicId)));
        }
    }
}
DynamicIdDirective.decorators = [
    { type: Directive, args: [{
                selector: "[dynamicId]"
            },] }
];
/** @nocollapse */
DynamicIdDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
DynamicIdDirective.propDecorators = {
    dynamicId: [{ type: Input }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DynamicListDirective {
    /**
     * @param {?} elementRef
     * @param {?} renderer
     */
    constructor(elementRef, renderer) {
        this.elementRef = elementRef;
        this.renderer = renderer;
    }
    /**
     * @return {?}
     */
    ngAfterViewInit() {
        if (isString(this.listId)) {
            this.renderer.setAttribute(this.elementRef.nativeElement, "list", (/** @type {?} */ (this.listId)));
        }
    }
}
DynamicListDirective.decorators = [
    { type: Directive, args: [{
                selector: "[dynamicList]"
            },] }
];
/** @nocollapse */
DynamicListDirective.ctorParameters = () => [
    { type: ElementRef },
    { type: Renderer2 }
];
DynamicListDirective.propDecorators = {
    listId: [{ type: Input, args: ["dynamicList",] }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @enum {string} */
const DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT = {
    Start: "START", End: "END",
};
class DynamicTemplateDirective {
    /**
     * @param {?} templateRef
     */
    constructor(templateRef) {
        this.templateRef = templateRef;
        this.align = DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT.End;
        this.as = null;
    }
}
DynamicTemplateDirective.decorators = [
    { type: Directive, args: [{
                selector: "ng-template[modelId],ng-template[modelType]"
            },] }
];
/** @nocollapse */
DynamicTemplateDirective.ctorParameters = () => [
    { type: TemplateRef }
];
DynamicTemplateDirective.propDecorators = {
    align: [{ type: Input }],
    as: [{ type: Input }],
    index: [{ type: Input }],
    modelId: [{ type: Input }],
    modelType: [{ type: Input }]
};

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
class DynamicDateControlModel extends DynamicFormValueControlModel {
    /**
     * @protected
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.format = config.format || null;
        this.max = config.max || null;
        this.min = config.min || null;
        this.placeholder = config.placeholder || null;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicDateControlModel.prototype, "format", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicDateControlModel.prototype, "max", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicDateControlModel.prototype, "min", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicDateControlModel.prototype, "placeholder", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @abstract
 */
class DynamicFileControlModel extends DynamicFormValueControlModel {
    /**
     * @protected
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.multiple = isBoolean(config.multiple) ? config.multiple : false;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicFileControlModel.prototype, "multiple", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/**
 * @template T
 */
class DynamicFormOption {
    /**
     * @param {?} config
     */
    constructor(config) {
        this.disabled = isBoolean(config.disabled) ? config.disabled : false;
        this.label = config.label || null;
        this.value = config.value;
    }
    /**
     * @return {?}
     */
    get text() {
        return this.label;
    }
    /**
     * @param {?} text
     * @return {?}
     */
    set text(text) {
        this.label = text;
    }
    /**
     * @return {?}
     */
    toJSON() {
        return serialize(this);
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicFormOption.prototype, "disabled", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFormOption.prototype, "label", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicFormOption.prototype, "value", void 0);
/**
 * @abstract
 * @template T
 */
class DynamicOptionControlModel extends DynamicFormValueControlModel {
    /**
     * @protected
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this._options = [];
        this.options = config.options;
    }
    /**
     * @private
     * @return {?}
     */
    updateOptions$() {
        this.options$ = of(this.options);
    }
    /**
     * @param {?} options
     * @return {?}
     */
    set options(options) {
        if (Array.isArray(options)) {
            this._options = ((/** @type {?} */ (options))).map((/**
             * @param {?} optionConfig
             * @return {?}
             */
            optionConfig => {
                return new DynamicFormOption(optionConfig);
            }));
            this.updateOptions$();
        }
        else if (isObservable(options)) {
            this.options$ = ((/** @type {?} */ (options))).pipe(map((/**
             * @param {?} optionsConfig
             * @return {?}
             */
            optionsConfig => {
                /** @type {?} */
                let options = optionsConfig.map((/**
                 * @param {?} optionConfig
                 * @return {?}
                 */
                optionConfig => new DynamicFormOption(optionConfig)));
                this._options = options;
                return options;
            })));
        }
        else {
            this.updateOptions$();
        }
    }
    /**
     * @return {?}
     */
    get options() {
        return this._options;
    }
    /**
     * @param {?} optionConfig
     * @return {?}
     */
    add(optionConfig) {
        return this.insert(this.options.length, optionConfig);
    }
    /**
     * @param {?} index
     * @return {?}
     */
    get(index) {
        return this.options[index];
    }
    /**
     * @param {?} index
     * @param {?} optionConfig
     * @return {?}
     */
    insert(index, optionConfig) {
        /** @type {?} */
        let option = new DynamicFormOption(optionConfig);
        this.options.splice(index, 0, option);
        this.updateOptions$();
        return option;
    }
    /**
     * @param {...?} indices
     * @return {?}
     */
    remove(...indices) {
        indices.forEach((/**
         * @param {?} index
         * @return {?}
         */
        index => this.options.splice(index, 1)));
        this.updateOptions$();
    }
}
__decorate([
    serializable("options"),
    __metadata("design:type", Array)
], DynamicOptionControlModel.prototype, "_options", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_GROUP = "GROUP";
class DynamicFormGroupModel extends DynamicFormControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.group = [];
        this.type = DYNAMIC_FORM_CONTROL_TYPE_GROUP;
        this.group = Array.isArray(config.group) ? config.group : [];
        this.legend = config.legend || null;
    }
    /**
     * @param {?} index
     * @return {?}
     */
    get(index) {
        return this.group[index];
    }
    /**
     * @param {?} index
     * @param {?} controlModel
     * @return {?}
     */
    set(index, controlModel) {
        this.group[index] = controlModel;
    }
    /**
     * @param {?} controlModel
     * @return {?}
     */
    add(controlModel) {
        this.group.push(controlModel);
    }
    /**
     * @param {?} index
     * @param {?} controlModel
     * @return {?}
     */
    insert(index, controlModel) {
        this.group.splice(index, 0, controlModel);
    }
    /**
     * @param {?} index
     * @param {?} step
     * @return {?}
     */
    move(index, step) {
        this.group.splice(index + step, 0, ...this.group.splice(index, 1));
    }
    /**
     * @param {?} index
     * @return {?}
     */
    remove(index) {
        this.group.splice(index, 1);
    }
    /**
     * @return {?}
     */
    size() {
        return this.group.length;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Array)
], DynamicFormGroupModel.prototype, "group", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFormGroupModel.prototype, "legend", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFormGroupModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP = "CHECKBOX_GROUP";
class DynamicCheckboxGroupModel extends DynamicFormGroupModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP;
    }
    /**
     * @param {...?} indices
     * @return {?}
     */
    check(...indices) {
        indices.forEach((/**
         * @param {?} index
         * @return {?}
         */
        index => this.group[index].checked = true));
    }
    /**
     * @param {...?} indices
     * @return {?}
     */
    uncheck(...indices) {
        indices.forEach((/**
         * @param {?} index
         * @return {?}
         */
        index => this.group[index].checked = false));
    }
    /**
     * @return {?}
     */
    checkAll() {
        this.group.forEach((/**
         * @param {?} model
         * @return {?}
         */
        model => model.checked = true));
    }
    /**
     * @return {?}
     */
    uncheckAll() {
        this.group.forEach((/**
         * @param {?} model
         * @return {?}
         */
        model => model.checked = false));
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Array)
], DynamicCheckboxGroupModel.prototype, "group", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicCheckboxGroupModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_COLORPICKER = "COLORPICKER";
class DynamicColorPickerModel extends DynamicFormValueControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_COLORPICKER;
        this.format = isString(config.format) ? config.format : null;
        this.inline = isBoolean(config.inline) ? config.inline : false;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicColorPickerModel.prototype, "format", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicColorPickerModel.prototype, "inline", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicColorPickerModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_DATEPICKER = "DATEPICKER";
class DynamicDatePickerModel extends DynamicDateControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_DATEPICKER;
        this.autoFocus = isBoolean(config.autoFocus) ? config.autoFocus : false;
        this.focusedDate = config.focusedDate || null;
        this.inline = isBoolean(config.inline) ? config.inline : false;
        this.prefix = config.prefix || null;
        this.readOnly = isBoolean(config.readOnly) ? config.readOnly : false;
        this.toggleIcon = isString(config.toggleIcon) ? config.toggleIcon : null;
        this.toggleLabel = isString(config.toggleLabel) ? config.toggleLabel : null;
        this.suffix = config.suffix || null;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicDatePickerModel.prototype, "autoFocus", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Object)
], DynamicDatePickerModel.prototype, "focusedDate", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicDatePickerModel.prototype, "inline", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicDatePickerModel.prototype, "prefix", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicDatePickerModel.prototype, "readOnly", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicDatePickerModel.prototype, "suffix", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicDatePickerModel.prototype, "toggleIcon", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicDatePickerModel.prototype, "toggleLabel", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicDatePickerModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_EDITOR = "EDITOR";
class DynamicEditorModel extends DynamicInputControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_EDITOR;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicEditorModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_FILE_UPLOAD = "FILE_UPLOAD";
class DynamicFileUploadModel extends DynamicFileControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_FILE_UPLOAD;
        this.accept = Array.isArray(config.accept) ? config.accept : null;
        this.autoUpload = isBoolean(config.autoUpload) ? config.autoUpload : true;
        this.maxSize = isNumber(config.maxSize) ? config.maxSize : null;
        this.minSize = isNumber(config.minSize) ? config.minSize : null;
        this.removeUrl = config.removeUrl || null;
        this.showFileList = isBoolean(config.showFileList) ? config.showFileList : true;
        this.url = config.url || null;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Array)
], DynamicFileUploadModel.prototype, "accept", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicFileUploadModel.prototype, "autoUpload", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicFileUploadModel.prototype, "maxSize", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicFileUploadModel.prototype, "minSize", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFileUploadModel.prototype, "removeUrl", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicFileUploadModel.prototype, "showFileList", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFileUploadModel.prototype, "url", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicFileUploadModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_RADIO_GROUP = "RADIO_GROUP";
/**
 * @template T
 */
class DynamicRadioGroupModel extends DynamicOptionControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_RADIO_GROUP;
        this.legend = config.legend || null;
    }
    /**
     * @param {?} index
     * @return {?}
     */
    select(index) {
        this.value = this.get(index).value;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicRadioGroupModel.prototype, "legend", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicRadioGroupModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_RATING = "RATING";
class DynamicRatingModel extends DynamicFormValueControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_RATING;
        this.max = isNumber(config.max) ? config.max : 10;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicRatingModel.prototype, "max", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicRatingModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_SELECT = "SELECT";
/**
 * @template T
 */
class DynamicSelectModel extends DynamicOptionControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_SELECT;
        this.compareWithFn = isFunction(config.compareWithFn) ? config.compareWithFn : ɵlooseIdentical;
        this.filterable = isBoolean(config.filterable) ? config.filterable : false;
        this.multiple = isBoolean(config.multiple) ? config.multiple : false;
        this.placeholder = config.placeholder || "";
        this.prefix = config.prefix || null;
        this.suffix = config.suffix || null;
    }
    /**
     * @param {...?} indices
     * @return {?}
     */
    select(...indices) {
        /** @type {?} */
        const value = this.multiple ? indices.map((/**
         * @param {?} index
         * @return {?}
         */
        index => this.get(index).value)) : this.get(indices[0]).value;
        this.value = value;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicSelectModel.prototype, "filterable", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicSelectModel.prototype, "multiple", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicSelectModel.prototype, "placeholder", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicSelectModel.prototype, "prefix", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicSelectModel.prototype, "suffix", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicSelectModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_SLIDER = "SLIDER";
class DynamicSliderModel extends DynamicFormValueControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_SLIDER;
        this.max = isNumber(config.max) ? config.max : 10;
        this.min = isNumber(config.min) ? config.min : 0;
        this.step = isNumber(config.step) ? config.step : 1;
        this.vertical = isBoolean(config.vertical) ? config.vertical : false;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicSliderModel.prototype, "max", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicSliderModel.prototype, "min", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicSliderModel.prototype, "step", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicSliderModel.prototype, "vertical", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicSliderModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_SWITCH = "SWITCH";
class DynamicSwitchModel extends DynamicCheckControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_SWITCH;
        this.offLabel = config.offLabel || null;
        this.onLabel = config.onLabel || null;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicSwitchModel.prototype, "offLabel", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicSwitchModel.prototype, "onLabel", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicSwitchModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_TEXTAREA = "TEXTAREA";
/** @type {?} */
const DYNAMIC_FORM_TEXTAREA_WRAP_HARD = "hard";
/** @type {?} */
const DYNAMIC_FORM_TEXTAREA_WRAP_SOFT = "soft";
class DynamicTextAreaModel extends DynamicInputControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_TEXTAREA;
        this.cols = isNumber(config.cols) ? config.cols : 20;
        this.rows = isNumber(config.rows) ? config.rows : 2;
        this.wrap = config.wrap || DYNAMIC_FORM_TEXTAREA_WRAP_SOFT;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicTextAreaModel.prototype, "cols", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Number)
], DynamicTextAreaModel.prototype, "rows", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicTextAreaModel.prototype, "wrap", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicTextAreaModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_TYPE_TIMEPICKER = "TIMEPICKER";
class DynamicTimePickerModel extends DynamicDateControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_TIMEPICKER;
        this.meridian = isBoolean(config.meridian) ? config.meridian : false;
        this.showSeconds = isBoolean(config.showSeconds) ? config.showSeconds : false;
    }
}
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicTimePickerModel.prototype, "meridian", void 0);
__decorate([
    serializable(),
    __metadata("design:type", Boolean)
], DynamicTimePickerModel.prototype, "showSeconds", void 0);
__decorate([
    serializable(),
    __metadata("design:type", String)
], DynamicTimePickerModel.prototype, "type", void 0);

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_VALIDATORS = new InjectionToken("DYNAMIC_VALIDATORS");

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DynamicFormValidationService {
    /**
     * @param {?} NG_VALIDATORS
     * @param {?} NG_ASYNC_VALIDATORS
     * @param {?} DYNAMIC_VALIDATORS
     */
    constructor(NG_VALIDATORS, NG_ASYNC_VALIDATORS, DYNAMIC_VALIDATORS) {
        this.NG_VALIDATORS = NG_VALIDATORS;
        this.NG_ASYNC_VALIDATORS = NG_ASYNC_VALIDATORS;
        this.DYNAMIC_VALIDATORS = DYNAMIC_VALIDATORS;
    }
    /**
     * @private
     * @param {?} validatorName
     * @param {?=} validatorArgs
     * @param {?=} validatorsToken
     * @return {?}
     */
    getValidatorFn(validatorName, validatorArgs = null, validatorsToken = this.NG_VALIDATORS) {
        /** @type {?} */
        let validatorFn;
        if (Validators.hasOwnProperty(validatorName)) { // Built-in Angular Validators
            validatorFn = ((/** @type {?} */ (Validators)))[validatorName];
        }
        else { // Custom Validators
            if (this.DYNAMIC_VALIDATORS && this.DYNAMIC_VALIDATORS.has(validatorName)) {
                validatorFn = this.DYNAMIC_VALIDATORS.get(validatorName);
            }
            else if (validatorsToken) {
                validatorFn = validatorsToken.find((/**
                 * @param {?} validatorFn
                 * @return {?}
                 */
                validatorFn => validatorFn.name === validatorName));
            }
        }
        if (validatorFn === undefined) { // throw when no validator could be resolved
            throw new Error(`validator "${validatorName}" is not provided via NG_VALIDATORS, NG_ASYNC_VALIDATORS or DYNAMIC_FORM_VALIDATORS`);
        }
        if (validatorArgs !== null) {
            return ((/** @type {?} */ (validatorFn)))(validatorArgs);
        }
        return (/** @type {?} */ (validatorFn));
    }
    /**
     * @private
     * @param {?} validatorsConfig
     * @param {?=} validatorsToken
     * @return {?}
     */
    getValidatorFns(validatorsConfig, validatorsToken = this.NG_VALIDATORS) {
        /** @type {?} */
        let validatorFns = [];
        if (isObject(validatorsConfig)) {
            validatorFns = Object.keys(validatorsConfig).map((/**
             * @param {?} validatorConfigKey
             * @return {?}
             */
            validatorConfigKey => {
                /** @type {?} */
                let validatorConfigValue = ((/** @type {?} */ (validatorsConfig)))[validatorConfigKey];
                if (this.isValidatorDescriptor(validatorConfigValue)) {
                    /** @type {?} */
                    let descriptor = (/** @type {?} */ (validatorConfigValue));
                    return this.getValidatorFn(descriptor.name, descriptor.args, validatorsToken);
                }
                return this.getValidatorFn(validatorConfigKey, validatorConfigValue, validatorsToken);
            }));
        }
        return validatorFns;
    }
    /**
     * @private
     * @param {?} template
     * @param {?} model
     * @param {?=} error
     * @return {?}
     */
    parseErrorMessageConfig(template, model, error = null) {
        return template.replace(/{{\s*(.+?)\s*}}/mg, (/**
         * @param {?} _match
         * @param {?} expression
         * @return {?}
         */
        (_match, expression) => {
            /** @type {?} */
            let propertySource = model;
            /** @type {?} */
            let propertyName = expression;
            if (expression.indexOf("validator.") >= 0 && error) {
                propertySource = error;
                propertyName = expression.replace("validator.", "");
            }
            return propertySource[propertyName] !== null && propertySource[propertyName] !== undefined ?
                propertySource[propertyName] : null;
        }));
    }
    /**
     * @param {?} validatorName
     * @param {?=} validatorArgs
     * @return {?}
     */
    getValidator(validatorName, validatorArgs = null) {
        return (/** @type {?} */ (this.getValidatorFn(validatorName, validatorArgs)));
    }
    /**
     * @param {?} validatorName
     * @param {?=} validatorArgs
     * @return {?}
     */
    getAsyncValidator(validatorName, validatorArgs = null) {
        return (/** @type {?} */ (this.getValidatorFn(validatorName, validatorArgs, this.NG_ASYNC_VALIDATORS)));
    }
    /**
     * @param {?} validatorsConfig
     * @return {?}
     */
    getValidators(validatorsConfig) {
        return (/** @type {?} */ (this.getValidatorFns(validatorsConfig)));
    }
    /**
     * @param {?} asyncValidatorsConfig
     * @return {?}
     */
    getAsyncValidators(asyncValidatorsConfig) {
        return (/** @type {?} */ (this.getValidatorFns(asyncValidatorsConfig, this.NG_ASYNC_VALIDATORS)));
    }
    /**
     * @param {?} validatorsConfig
     * @param {?} control
     * @param {?} model
     * @return {?}
     */
    updateValidators(validatorsConfig, control, model) {
        model.validators = validatorsConfig;
        if (validatorsConfig === null) {
            control.clearValidators();
        }
        else {
            control.setValidators(this.getValidators(validatorsConfig));
        }
        control.updateValueAndValidity();
    }
    /**
     * @param {?} asyncValidatorsConfig
     * @param {?} control
     * @param {?} model
     * @return {?}
     */
    updateAsyncValidators(asyncValidatorsConfig, control, model) {
        model.asyncValidators = asyncValidatorsConfig;
        if (asyncValidatorsConfig === null) {
            control.clearAsyncValidators();
        }
        else {
            control.setAsyncValidators(this.getAsyncValidators(asyncValidatorsConfig));
        }
        control.updateValueAndValidity();
    }
    /**
     * @param {?} control
     * @param {?} model
     * @return {?}
     */
    createErrorMessages(control, model) {
        /** @type {?} */
        let messages = [];
        if (model.hasErrorMessages) {
            /** @type {?} */
            let messagesConfig = (/** @type {?} */ (model.errorMessages));
            Object.keys(control.errors || {}).forEach((/**
             * @param {?} validationErrorKey
             * @return {?}
             */
            validationErrorKey => {
                /** @type {?} */
                let messageKey = validationErrorKey;
                if (validationErrorKey === "minlength" || validationErrorKey === "maxlength") {
                    messageKey = messageKey.replace("length", "Length");
                }
                if (messagesConfig.hasOwnProperty(messageKey)) {
                    /** @type {?} */
                    let validationError = control.getError(validationErrorKey);
                    /** @type {?} */
                    let messageTemplate = (/** @type {?} */ (messagesConfig[messageKey]));
                    messages.push(this.parseErrorMessageConfig(messageTemplate, model, validationError));
                }
            }));
        }
        return messages;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    isFormHook(value) {
        return isString(value) && ["blur", "change", "submit"].indexOf(value) !== -1;
    }
    /**
     * @param {?} value
     * @return {?}
     */
    isValidatorDescriptor(value) {
        if (isObject(value)) {
            return value.hasOwnProperty("name") && value.hasOwnProperty("args");
        }
        return false;
    }
}
DynamicFormValidationService.decorators = [
    { type: Injectable, args: [{
                providedIn: "root"
            },] }
];
/** @nocollapse */
DynamicFormValidationService.ctorParameters = () => [
    { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [NG_VALIDATORS,] }] },
    { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [NG_ASYNC_VALIDATORS,] }] },
    { type: Map, decorators: [{ type: Optional }, { type: Inject, args: [DYNAMIC_VALIDATORS,] }] }
];
/** @nocollapse */ DynamicFormValidationService.ngInjectableDef = ɵɵdefineInjectable({ factory: function DynamicFormValidationService_Factory() { return new DynamicFormValidationService(ɵɵinject(NG_VALIDATORS, 8), ɵɵinject(NG_ASYNC_VALIDATORS, 8), ɵɵinject(DYNAMIC_VALIDATORS, 8)); }, token: DynamicFormValidationService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DynamicFormService {
    /**
     * @param {?} validationService
     */
    constructor(validationService) {
        this.validationService = validationService;
    }
    /**
     * @private
     * @param {?=} validatorsConfig
     * @param {?=} asyncValidatorsConfig
     * @param {?=} updateOn
     * @return {?}
     */
    createAbstractControlOptions(validatorsConfig = null, asyncValidatorsConfig = null, updateOn = null) {
        return {
            asyncValidators: asyncValidatorsConfig !== null ? this.validationService.getAsyncValidators(asyncValidatorsConfig) : null,
            validators: validatorsConfig !== null ? this.validationService.getValidators(validatorsConfig) : null,
            updateOn: updateOn !== null && this.validationService.isFormHook(updateOn) ? updateOn : "change"
        };
    }
    /**
     * @param {?} formArrayModel
     * @return {?}
     */
    createFormArray(formArrayModel) {
        /** @type {?} */
        let controls = [];
        /** @type {?} */
        let options = this.createAbstractControlOptions(formArrayModel.validators, formArrayModel.asyncValidators, formArrayModel.updateOn);
        for (let index = 0; index < formArrayModel.size; index++) {
            /** @type {?} */
            let groupModel = formArrayModel.get(index);
            /** @type {?} */
            let groupOptions = this.createAbstractControlOptions(formArrayModel.groupValidators, formArrayModel.groupAsyncValidators, formArrayModel.updateOn);
            controls.push(this.createFormGroup(groupModel.group, groupOptions, groupModel));
        }
        return new FormArray(controls, options);
    }
    /**
     * @param {?} formModel
     * @param {?=} options
     * @param {?=} parent
     * @return {?}
     */
    createFormGroup(formModel, options = null, parent = null) {
        /** @type {?} */
        let controls = {};
        formModel.forEach((/**
         * @param {?} model
         * @return {?}
         */
        model => {
            model.parent = parent;
            switch (model.type) {
                case DYNAMIC_FORM_CONTROL_TYPE_ARRAY:
                    controls[model.id] = this.createFormArray((/** @type {?} */ (model)));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_GROUP:
                case DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP:
                    /** @type {?} */
                    let groupModel = (/** @type {?} */ (model));
                    /** @type {?} */
                    let groupOptions = this.createAbstractControlOptions(groupModel.validators, groupModel.asyncValidators, groupModel.updateOn);
                    controls[model.id] = this.createFormGroup(groupModel.group, groupOptions, groupModel);
                    break;
                default:
                    /** @type {?} */
                    let controlModel = (/** @type {?} */ (model));
                    /** @type {?} */
                    let controlState = { value: controlModel.value, disabled: controlModel.disabled };
                    /** @type {?} */
                    let controlOptions = this.createAbstractControlOptions(controlModel.validators, controlModel.asyncValidators, controlModel.updateOn);
                    controls[model.id] = new FormControl(controlState, controlOptions);
            }
        }));
        return new FormGroup(controls, options);
    }
    /**
     * @param {?} model
     * @return {?}
     */
    getPathSegment(model) {
        return model instanceof DynamicFormArrayGroupModel ? model.index.toString() : ((/** @type {?} */ (model))).id;
    }
    /**
     * @param {?} model
     * @param {?=} join
     * @return {?}
     */
    getPath(model, join = false) {
        /** @type {?} */
        let path = [this.getPathSegment(model)];
        /** @type {?} */
        let parent = model.parent;
        while (parent) {
            path.unshift(this.getPathSegment(parent));
            parent = parent.parent;
        }
        return join ? path.join(".") : path;
    }
    /**
     * @param {?} formGroup
     * @param {?} formModel
     * @param {...?} models
     * @return {?}
     */
    addFormGroupControl(formGroup, formModel, ...models) {
        if (formModel instanceof DynamicFormGroupModel) {
            this.insertFormGroupControl(formModel.size(), formGroup, formModel, ...models);
        }
        else {
            /** @type {?} */
            let _formModel = (/** @type {?} */ (formModel));
            this.insertFormGroupControl(_formModel.length, formGroup, _formModel, ...models);
        }
    }
    /**
     * @param {?} index
     * @param {?} step
     * @param {?} formModel
     * @return {?}
     */
    moveFormGroupControl(index, step, formModel) {
        if (formModel instanceof DynamicFormGroupModel) {
            formModel.move(index, step);
        }
        else {
            /** @type {?} */
            let _formModel = (/** @type {?} */ (formModel));
            _formModel.splice(index + step, 0, ..._formModel.splice(index, 1));
        }
    }
    /**
     * @param {?} index
     * @param {?} formGroup
     * @param {?} formModel
     * @param {...?} models
     * @return {?}
     */
    insertFormGroupControl(index, formGroup, formModel, ...models) {
        /** @type {?} */
        let parent = formModel instanceof DynamicFormGroupModel ? formModel : null;
        /** @type {?} */
        let controls = this.createFormGroup(models, null, parent).controls;
        Object.keys(controls).forEach((/**
         * @param {?} controlName
         * @param {?} idx
         * @return {?}
         */
        (controlName, idx) => {
            /** @type {?} */
            let controlModel = models[idx];
            if (formModel instanceof DynamicFormGroupModel) {
                formModel.insert(index, controlModel);
            }
            else {
                ((/** @type {?} */ (formModel))).splice(index, 0, controlModel);
            }
            formGroup.addControl(controlName, controls[controlName]);
        }));
    }
    /**
     * @param {?} index
     * @param {?} formGroup
     * @param {?} formModel
     * @return {?}
     */
    removeFormGroupControl(index, formGroup, formModel) {
        if (formModel instanceof DynamicFormGroupModel) {
            formGroup.removeControl(formModel.get(index).id);
            formModel.remove(index);
        }
        else {
            formGroup.removeControl(formModel[index].id);
            ((/** @type {?} */ (formModel))).splice(index, 1);
        }
    }
    /**
     * @param {?} formArray
     * @param {?} formArrayModel
     * @return {?}
     */
    addFormArrayGroup(formArray, formArrayModel) {
        /** @type {?} */
        let groupModel = formArrayModel.addGroup();
        formArray.push(this.createFormGroup(groupModel.group, null, groupModel));
    }
    /**
     * @param {?} index
     * @param {?} formArray
     * @param {?} formArrayModel
     * @return {?}
     */
    insertFormArrayGroup(index, formArray, formArrayModel) {
        /** @type {?} */
        let groupModel = formArrayModel.insertGroup(index);
        formArray.insert(index, this.createFormGroup(groupModel.group, null, groupModel));
    }
    /**
     * @param {?} index
     * @param {?} step
     * @param {?} formArray
     * @param {?} formArrayModel
     * @return {?}
     */
    moveFormArrayGroup(index, step, formArray, formArrayModel) {
        /** @type {?} */
        let newIndex = index + step;
        /** @type {?} */
        let moveUp = step >= 0;
        if ((index >= 0 && index < formArrayModel.size) && (newIndex >= 0 && newIndex < formArrayModel.size)) {
            /** @type {?} */
            let movingGroups = [];
            for (let i = moveUp ? index : newIndex; i <= (moveUp ? newIndex : index); i++) {
                movingGroups.push(formArray.at(i));
            }
            movingGroups.forEach((/**
             * @param {?} formControl
             * @param {?} idx
             * @return {?}
             */
            (formControl, idx) => {
                /** @type {?} */
                let position;
                if (moveUp) {
                    position = idx === 0 ? newIndex : index + idx - 1;
                }
                else {
                    position = idx === movingGroups.length - 1 ? newIndex : newIndex + idx + 1;
                }
                formArray.setControl(position, formControl);
            }));
            formArrayModel.moveGroup(index, step);
        }
        else {
            throw new Error(`form array group cannot be moved due to index or new index being out of bounds`);
        }
    }
    /**
     * @param {?} index
     * @param {?} formArray
     * @param {?} formArrayModel
     * @return {?}
     */
    removeFormArrayGroup(index, formArray, formArrayModel) {
        formArray.removeAt(index);
        formArrayModel.removeGroup(index);
    }
    /**
     * @param {?} formArray
     * @param {?} formArrayModel
     * @return {?}
     */
    clearFormArray(formArray, formArrayModel) {
        formArray.clear();
        formArrayModel.clear();
    }
    /**
     * @param {?} id
     * @param {?} formModel
     * @return {?}
     */
    findById(id, formModel) {
        /** @type {?} */
        let result = null;
        /** @type {?} */
        let findByIdFn = (/**
         * @param {?} id
         * @param {?} groupModel
         * @return {?}
         */
        (id, groupModel) => {
            for (let controlModel of groupModel) {
                if (controlModel.id === id) {
                    result = controlModel;
                    break;
                }
                if (controlModel instanceof DynamicFormGroupModel) {
                    findByIdFn(id, ((/** @type {?} */ (controlModel))).group);
                }
            }
        });
        findByIdFn(id, formModel);
        return result;
    }
    /**
     * @param {?} id
     * @param {?} formModel
     * @return {?}
     */
    findModelById(id, formModel) {
        return this.findById(id, formModel);
    }
    /**
     * @param {?} model
     * @param {?} group
     * @return {?}
     */
    findControlByModel(model, group) {
        return group.root.get(this.getPath(model, true));
    }
    /**
     * @param {?} json
     * @return {?}
     */
    fromJSON(json) {
        /** @type {?} */
        let formModelJSON = isString(json) ? JSON.parse(json, parseReviver) : json;
        /** @type {?} */
        let formModel = [];
        formModelJSON.forEach((/**
         * @param {?} model
         * @return {?}
         */
        (model) => {
            /** @type {?} */
            let layout = model.layout || null;
            switch (model.type) {
                case DYNAMIC_FORM_CONTROL_TYPE_ARRAY:
                    /** @type {?} */
                    let formArrayModel = (/** @type {?} */ (model));
                    if (Array.isArray(formArrayModel.groups)) {
                        formArrayModel.groups.forEach((/**
                         * @param {?} groupModel
                         * @return {?}
                         */
                        (groupModel) => {
                            groupModel.group = (/** @type {?} */ (this.fromJSON(groupModel.group)));
                        }));
                    }
                    formArrayModel.groupFactory = (/**
                     * @return {?}
                     */
                    () => {
                        return this.fromJSON(formArrayModel.groupPrototype);
                    });
                    formModel.push(new DynamicFormArrayModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX:
                    formModel.push(new DynamicCheckboxModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP:
                    model.group = (/** @type {?} */ (this.fromJSON(model.group)));
                    formModel.push(new DynamicCheckboxGroupModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_COLORPICKER:
                    formModel.push(new DynamicColorPickerModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_DATEPICKER:
                    formModel.push(new DynamicDatePickerModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_EDITOR:
                    formModel.push(new DynamicEditorModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_FILE_UPLOAD:
                    model.value = null;
                    formModel.push(new DynamicFileUploadModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_GROUP:
                    model.group = this.fromJSON(model.group);
                    formModel.push(new DynamicFormGroupModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_INPUT:
                    /** @type {?} */
                    let inputModel = (/** @type {?} */ (model));
                    if (inputModel.mask !== null) {
                        if (!(inputModel.mask instanceof Function)) {
                            inputModel.mask = maskFromString((/** @type {?} */ (inputModel.mask)));
                        }
                    }
                    formModel.push(new DynamicInputModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_RADIO_GROUP:
                    formModel.push(new DynamicRadioGroupModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_RATING:
                    formModel.push(new DynamicRatingModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_SELECT:
                    formModel.push(new DynamicSelectModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_SLIDER:
                    formModel.push(new DynamicSliderModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_SWITCH:
                    formModel.push(new DynamicSwitchModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_TEXTAREA:
                    formModel.push(new DynamicTextAreaModel(model, layout));
                    break;
                case DYNAMIC_FORM_CONTROL_TYPE_TIMEPICKER:
                    formModel.push(new DynamicTimePickerModel(model, layout));
                    break;
                default:
                    throw new Error(`unknown form control model type defined on JSON object with id "${model.id}"`);
            }
        }));
        return formModel;
    }
}
DynamicFormService.decorators = [
    { type: Injectable, args: [{
                providedIn: "root"
            },] }
];
/** @nocollapse */
DynamicFormService.ctorParameters = () => [
    { type: DynamicFormValidationService }
];
/** @nocollapse */ DynamicFormService.ngInjectableDef = ɵɵdefineInjectable({ factory: function DynamicFormService_Factory() { return new DynamicFormService(ɵɵinject(DynamicFormValidationService)); }, token: DynamicFormService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const DYNAMIC_FORM_CONTROL_MAP_FN = new InjectionToken("DYNAMIC_FORM_CONTROL_MAP_FN");
class DynamicFormComponentService {
    /**
     * @param {?} DYNAMIC_FORM_CONTROL_MAP_FN
     */
    constructor(DYNAMIC_FORM_CONTROL_MAP_FN) {
        this.DYNAMIC_FORM_CONTROL_MAP_FN = DYNAMIC_FORM_CONTROL_MAP_FN;
        this.componentRefs = {};
        this.DYNAMIC_FORM_CONTROL_MAP_FN = (/** @type {?} */ (DYNAMIC_FORM_CONTROL_MAP_FN));
    }
    /**
     * @param {?} modelId
     * @param {?=} index
     * @return {?}
     */
    getFormControlRef(modelId, index) {
        /** @type {?} */
        const ref = this.componentRefs[modelId];
        if (isNumber(index)) {
            return Array.isArray(ref) ? ref[index] : undefined;
        }
        else {
            return (/** @type {?} */ (ref));
        }
    }
    /**
     * @param {?} model
     * @param {?} instance
     * @param {?=} index
     * @return {?}
     */
    registerFormControlRef(model, instance, index) {
        if (isNumber(index)) {
            /** @type {?} */
            const arrayRef = (/** @type {?} */ (this.componentRefs[model.id])) || [];
            if (Array.isArray(arrayRef)) {
                arrayRef.splice(index, 0, instance);
                this.componentRefs[model.id] = arrayRef;
            }
            else {
                console.warn(`registerFormControlRef is called with index for a non-array form control: ${model.id}`);
            }
        }
        else {
            this.componentRefs[model.id] = instance;
        }
    }
    /**
     * @param {?} modelId
     * @param {?=} index
     * @return {?}
     */
    unregisterFormControlRef(modelId, index) {
        /** @type {?} */
        const componentRef = this.componentRefs[modelId];
        if (isNumber(index)) {
            if (Array.isArray(componentRef) && componentRef[index] !== undefined) {
                componentRef.splice(index, 1);
            }
        }
        else if (componentRef !== undefined) {
            delete this.componentRefs[modelId];
        }
    }
    /**
     * @param {?} model
     * @return {?}
     */
    getCustomComponentType(model) {
        return isFunction(this.DYNAMIC_FORM_CONTROL_MAP_FN) ? this.DYNAMIC_FORM_CONTROL_MAP_FN(model) : null;
    }
}
DynamicFormComponentService.decorators = [
    { type: Injectable, args: [{
                providedIn: "root"
            },] }
];
/** @nocollapse */
DynamicFormComponentService.ctorParameters = () => [
    { type: undefined, decorators: [{ type: Inject, args: [DYNAMIC_FORM_CONTROL_MAP_FN,] }, { type: Optional }] }
];
/** @nocollapse */ DynamicFormComponentService.ngInjectableDef = ɵɵdefineInjectable({ factory: function DynamicFormComponentService_Factory() { return new DynamicFormComponentService(ɵɵinject(DYNAMIC_FORM_CONTROL_MAP_FN, 8)); }, token: DynamicFormComponentService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DynamicFormLayoutService {
    /**
     * @param {?} id
     * @param {?} formLayout
     * @return {?}
     */
    findById(id, formLayout) {
        if (isObject(formLayout)) {
            for (let key of Object.keys(formLayout)) {
                if (key === id) {
                    return formLayout[key];
                }
            }
        }
        return null;
    }
    /**
     * @param {?} model
     * @param {?} formLayout
     * @return {?}
     */
    findByModel(model, formLayout) {
        /** @type {?} */
        let controlLayout = null;
        if (isObject(formLayout)) {
            for (let key of Object.keys(formLayout)) {
                key.split(",").forEach((/**
                 * @param {?} substring
                 * @return {?}
                 */
                substring => {
                    /** @type {?} */
                    const selector = substring.trim();
                    if (selector === model.id || selector === model.type) {
                        controlLayout = formLayout[key];
                    }
                }));
            }
        }
        return controlLayout;
    }
    /**
     * @param {?} model
     * @param {?} templates
     * @return {?}
     */
    filterTemplatesByModel(model, templates) {
        /** @type {?} */
        const filterCallback = (/**
         * @param {?} template
         * @return {?}
         */
        (template) => {
            return template.modelId === model.id || template.modelType === model.type;
        });
        if (templates instanceof QueryList) {
            return templates.filter(filterCallback);
        }
        else if (Array.isArray(templates)) {
            return templates.filter(filterCallback);
        }
        return [];
    }
    /**
     * @param {?} model
     * @param {?} templates
     * @param {?} alignment
     * @return {?}
     */
    getAlignedTemplate(model, templates, alignment) {
        return this.filterTemplatesByModel(model, templates)
            .find((/**
         * @param {?} template
         * @return {?}
         */
        template => template.as === null && template.align === alignment));
    }
    /*
        getIndexedTemplates(model: DynamicFormControlModel, templates: DynamicFormControlTemplates): DynamicTemplateDirective[] | undefined {
            return this.filterTemplatesByModel(model, templates).filter(template => template.as === null);
        }
        */
    /**
     * @param {?} model
     * @param {?} templates
     * @return {?}
     */
    getStartTemplate(model, templates) {
        return this.getAlignedTemplate(model, templates, DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT.Start);
    }
    /**
     * @param {?} model
     * @param {?} templates
     * @return {?}
     */
    getEndTemplate(model, templates) {
        return this.getAlignedTemplate(model, templates, DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT.End);
    }
    /**
     * @param {?} layout
     * @param {?} context
     * @param {?} place
     * @return {?}
     */
    getClass(layout, context, place) {
        if (layout !== null && layout.hasOwnProperty(context)) {
            /** @type {?} */
            let config = (/** @type {?} */ (layout[context]));
            if (config.hasOwnProperty(place)) {
                return (/** @type {?} */ (config[place]));
            }
        }
        return "";
    }
    /**
     * @param {?} model
     * @return {?}
     */
    getElementId(model) {
        /** @type {?} */
        let id = model.id;
        /** @type {?} */
        let parent = model.parent;
        while (parent !== null) {
            if (parent instanceof DynamicFormArrayGroupModel) {
                id = `${parent.context.id}-${parent.index}-${model.id}`;
                break;
            }
            parent = parent.parent;
        }
        return id;
    }
}
DynamicFormLayoutService.decorators = [
    { type: Injectable, args: [{
                providedIn: "root"
            },] }
];
/** @nocollapse */ DynamicFormLayoutService.ngInjectableDef = ɵɵdefineInjectable({ factory: function DynamicFormLayoutService_Factory() { return new DynamicFormLayoutService(); }, token: DynamicFormLayoutService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
/** @type {?} */
const MATCH_DISABLED = "DISABLED";
/** @type {?} */
const MATCH_ENABLED = "ENABLED";
/** @type {?} */
const MATCH_HIDDEN = "HIDDEN";
/** @type {?} */
const MATCH_OPTIONAL = "OPTIONAL";
/** @type {?} */
const MATCH_REQUIRED = "REQUIRED";
/** @type {?} */
const MATCH_VISIBLE = "VISIBLE";
/** @type {?} */
const AND_OPERATOR = "AND";
/** @type {?} */
const OR_OPERATOR = "OR";
/** @type {?} */
const DYNAMIC_MATCHERS = new InjectionToken("DYNAMIC_MATCHERS");
/** @type {?} */
const DisabledMatcher = {
    match: MATCH_DISABLED,
    opposingMatch: MATCH_ENABLED,
    /**
     * @param {?} hasMatch
     * @param {?} model
     * @return {?}
     */
    onChange(hasMatch, model) {
        model.disabled = hasMatch;
    }
};
/** @type {?} */
const HiddenMatcher = {
    match: MATCH_HIDDEN,
    opposingMatch: MATCH_VISIBLE,
    /**
     * @param {?} hasMatch
     * @param {?} model
     * @return {?}
     */
    onChange(hasMatch, model) {
        model.hidden = hasMatch;
    }
};
/** @type {?} */
const RequiredMatcher = {
    match: MATCH_REQUIRED,
    opposingMatch: MATCH_OPTIONAL,
    /**
     * @param {?} hasMatch
     * @param {?} model
     * @param {?} control
     * @param {?} injector
     * @return {?}
     */
    onChange(hasMatch, model, control, injector) {
        /** @type {?} */
        let validatorsConfig = null;
        if (hasMatch) {
            validatorsConfig = isObject(model.validators) ? Object.assign({}, model.validators, { required: null }) : { required: null };
        }
        else {
            if (isObject(model.validators)) {
                delete model.validators["required"];
                validatorsConfig = Object.assign({}, model.validators);
            }
        }
        injector.get(DynamicFormValidationService).updateValidators(validatorsConfig, control, model);
    }
};
/** @type {?} */
const DISABLED_MATCHER = {
    provide: DYNAMIC_MATCHERS,
    useValue: DisabledMatcher,
    multi: true
};
/** @type {?} */
const HIDDEN_MATCHER = {
    provide: DYNAMIC_MATCHERS,
    useValue: HiddenMatcher,
    multi: true
};
/** @type {?} */
const REQUIRED_MATCHER = {
    provide: DYNAMIC_MATCHERS,
    useValue: RequiredMatcher,
    multi: true
};
/** @type {?} */
const DYNAMIC_MATCHER_PROVIDERS = [DISABLED_MATCHER, HIDDEN_MATCHER, REQUIRED_MATCHER];

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DynamicFormRelationService {
    /**
     * @param {?} DYNAMIC_MATCHERS
     * @param {?} injector
     */
    constructor(DYNAMIC_MATCHERS, injector) {
        this.DYNAMIC_MATCHERS = DYNAMIC_MATCHERS;
        this.injector = injector;
    }
    /**
     * @param {?} model
     * @param {?} group
     * @return {?}
     */
    getRelatedFormControls(model, group) {
        /** @type {?} */
        const conditionReducer = (/**
         * @param {?} controls
         * @param {?} condition
         * @return {?}
         */
        (controls, condition) => {
            /** @type {?} */
            const path = condition.rootPath || condition.id;
            if (!controls.hasOwnProperty(path)) {
                /** @type {?} */
                const control = condition.rootPath ? group.root.get(condition.rootPath) : group.get(condition.id);
                control instanceof FormControl ? controls[path] = control : console.warn(`No related form control with id ${condition.id} could be found`);
            }
            return controls;
        });
        /** @type {?} */
        const relationReducer = (/**
         * @param {?} controls
         * @param {?} relation
         * @return {?}
         */
        (controls, relation) => relation.when.reduce(conditionReducer, controls));
        return model.relations.reduce(relationReducer, {});
    }
    /**
     * @param {?} relations
     * @param {?} matcher
     * @return {?}
     */
    findRelationByMatcher(relations, matcher) {
        return relations.find((/**
         * @param {?} relation
         * @return {?}
         */
        relation => [matcher.match, matcher.opposingMatch].includes(relation.match)));
    }
    /**
     * @param {?} relation
     * @param {?} relatedFormControls
     * @param {?} matcher
     * @return {?}
     */
    matchesCondition(relation, relatedFormControls, matcher) {
        /** @type {?} */
        const operator = relation.operator || OR_OPERATOR;
        return relation.when.reduce((/**
         * @param {?} hasAlreadyMatched
         * @param {?} condition
         * @param {?} index
         * @return {?}
         */
        (hasAlreadyMatched, condition, index) => {
            /** @type {?} */
            const path = condition.rootPath || condition.id;
            /** @type {?} */
            let relatedFormControl;
            for (let [key, control] of Object.entries(relatedFormControls)) {
                if (key === path) {
                    relatedFormControl = control;
                    break;
                }
            }
            if (relatedFormControl && relation.match === matcher.match) {
                if (index > 0 && operator === AND_OPERATOR && !hasAlreadyMatched) {
                    return false;
                }
                if (index > 0 && operator === OR_OPERATOR && hasAlreadyMatched) {
                    return true;
                }
                return condition.value === relatedFormControl.value || condition.status === relatedFormControl.status;
            }
            if (relatedFormControl && relation.match === matcher.opposingMatch) {
                if (index > 0 && operator === AND_OPERATOR && hasAlreadyMatched) {
                    return true;
                }
                if (index > 0 && operator === OR_OPERATOR && !hasAlreadyMatched) {
                    return false;
                }
                return !(condition.value === relatedFormControl.value || condition.status === relatedFormControl.status);
            }
            return false;
        }), false);
    }
    /**
     * @param {?} model
     * @param {?} group
     * @param {?} control
     * @return {?}
     */
    subscribeRelations(model, group, control) {
        /** @type {?} */
        const relatedFormControls = this.getRelatedFormControls(model, group);
        /** @type {?} */
        const subscriptions = [];
        Object.values(relatedFormControls).forEach((/**
         * @param {?} relatedControl
         * @return {?}
         */
        relatedControl => {
            /** @type {?} */
            const valueChanges = relatedControl.valueChanges.pipe(startWith(relatedControl.value));
            /** @type {?} */
            const statusChanges = relatedControl.statusChanges.pipe(startWith(relatedControl.status));
            subscriptions.push(merge(valueChanges, statusChanges).subscribe((/**
             * @return {?}
             */
            () => {
                this.DYNAMIC_MATCHERS.forEach((/**
                 * @param {?} matcher
                 * @return {?}
                 */
                matcher => {
                    /** @type {?} */
                    const relation = this.findRelationByMatcher(model.relations, matcher);
                    if (relation !== undefined) {
                        /** @type {?} */
                        const hasMatch = this.matchesCondition(relation, relatedFormControls, matcher);
                        matcher.onChange(hasMatch, model, control, this.injector);
                    }
                }));
            })));
        }));
        return subscriptions;
    }
}
DynamicFormRelationService.decorators = [
    { type: Injectable, args: [{
                providedIn: "root"
            },] }
];
/** @nocollapse */
DynamicFormRelationService.ctorParameters = () => [
    { type: Array, decorators: [{ type: Optional }, { type: Inject, args: [DYNAMIC_MATCHERS,] }] },
    { type: Injector }
];
/** @nocollapse */ DynamicFormRelationService.ngInjectableDef = ɵɵdefineInjectable({ factory: function DynamicFormRelationService_Factory() { return new DynamicFormRelationService(ɵɵinject(DYNAMIC_MATCHERS, 8), ɵɵinject(INJECTOR)); }, token: DynamicFormRelationService, providedIn: "root" });

/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
class DynamicFormsCoreModule {
    /*@deprecated*/
    /**
     * @return {?}
     */
    static forRoot() {
        return {
            ngModule: DynamicFormsCoreModule,
            providers: [
                DynamicFormService,
                DynamicFormLayoutService,
                DynamicFormValidationService,
                DynamicFormComponentService,
                DynamicFormRelationService
            ]
        };
    }
}
DynamicFormsCoreModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    ReactiveFormsModule
                ],
                declarations: [
                    DynamicIdDirective,
                    DynamicListDirective,
                    DynamicTemplateDirective
                ],
                exports: [
                    DynamicIdDirective,
                    DynamicListDirective,
                    DynamicTemplateDirective
                ]
            },] }
];

export { AND_OPERATOR, AUTOCOMPLETE_OFF, AUTOCOMPLETE_ON, AUTOFILL_FIELDS, AUTOFILL_FIELDS_CONTACT, AUTOFILL_FIELD_ADDITIONAL_NAME, AUTOFILL_FIELD_ADDRESS_LEVEL_1, AUTOFILL_FIELD_ADDRESS_LEVEL_2, AUTOFILL_FIELD_ADDRESS_LEVEL_3, AUTOFILL_FIELD_ADDRESS_LEVEL_4, AUTOFILL_FIELD_ADDRESS_LINE_1, AUTOFILL_FIELD_ADDRESS_LINE_2, AUTOFILL_FIELD_ADDRESS_LINE_3, AUTOFILL_FIELD_BDAY, AUTOFILL_FIELD_BDAY_DAY, AUTOFILL_FIELD_BDAY_MONTH, AUTOFILL_FIELD_BDAY_YEAR, AUTOFILL_FIELD_CC_ADDITIONAL_NAME, AUTOFILL_FIELD_CC_CSC, AUTOFILL_FIELD_CC_EXP, AUTOFILL_FIELD_CC_EXP_MONTH, AUTOFILL_FIELD_CC_EXP_YEAR, AUTOFILL_FIELD_CC_FAMILY_NAME, AUTOFILL_FIELD_CC_GIVEN_NAME, AUTOFILL_FIELD_CC_NAME, AUTOFILL_FIELD_CC_NUMBER, AUTOFILL_FIELD_CC_TYPE, AUTOFILL_FIELD_COUNTRY, AUTOFILL_FIELD_COUNTRY_NAME, AUTOFILL_FIELD_CURRENT_PASSWORD, AUTOFILL_FIELD_EMAIL, AUTOFILL_FIELD_FAMILY_NAME, AUTOFILL_FIELD_GIVEN_NAME, AUTOFILL_FIELD_HONORIFIC_PREFIX, AUTOFILL_FIELD_HONORIFIC_SUFFIX, AUTOFILL_FIELD_IMPP, AUTOFILL_FIELD_LANGUAGE, AUTOFILL_FIELD_NAME, AUTOFILL_FIELD_NEW_PASSWORD, AUTOFILL_FIELD_NICKNAME, AUTOFILL_FIELD_ORGANIZATION, AUTOFILL_FIELD_ORGANIZATION_TITLE, AUTOFILL_FIELD_PHOTO, AUTOFILL_FIELD_POSTAL_CODE, AUTOFILL_FIELD_SEX, AUTOFILL_FIELD_STREET_ADDRESS, AUTOFILL_FIELD_TEL, AUTOFILL_FIELD_TEL_AREA_CODE, AUTOFILL_FIELD_TEL_COUNTRY_CODE, AUTOFILL_FIELD_TEL_LOCAL, AUTOFILL_FIELD_TEL_LOCAL_EXTENSION, AUTOFILL_FIELD_TEL_LOCAL_PREFIX, AUTOFILL_FIELD_TEL_LOCAL_SUFFIX, AUTOFILL_FIELD_TEL_NATIONAL, AUTOFILL_FIELD_TRANSACTION_AMOUNT, AUTOFILL_FIELD_TRANSACTION_CURRENCY, AUTOFILL_FIELD_URL, AUTOFILL_FIELD_USERNAME, AUTOFILL_TOKENS_ADDRESS, AUTOFILL_TOKENS_CONTACT, AUTOFILL_TOKEN_BILLING, AUTOFILL_TOKEN_FAX, AUTOFILL_TOKEN_HOME, AUTOFILL_TOKEN_MOBILE, AUTOFILL_TOKEN_PAGER, AUTOFILL_TOKEN_SHIPPING, AUTOFILL_TOKEN_WORK, DISABLED_MATCHER, DYNAMIC_FORM_CONTROL_INPUT_TYPE_COLOR, DYNAMIC_FORM_CONTROL_INPUT_TYPE_DATE, DYNAMIC_FORM_CONTROL_INPUT_TYPE_DATETIME_LOCAL, DYNAMIC_FORM_CONTROL_INPUT_TYPE_EMAIL, DYNAMIC_FORM_CONTROL_INPUT_TYPE_FILE, DYNAMIC_FORM_CONTROL_INPUT_TYPE_MONTH, DYNAMIC_FORM_CONTROL_INPUT_TYPE_NUMBER, DYNAMIC_FORM_CONTROL_INPUT_TYPE_PASSWORD, DYNAMIC_FORM_CONTROL_INPUT_TYPE_RANGE, DYNAMIC_FORM_CONTROL_INPUT_TYPE_SEARCH, DYNAMIC_FORM_CONTROL_INPUT_TYPE_TEL, DYNAMIC_FORM_CONTROL_INPUT_TYPE_TEXT, DYNAMIC_FORM_CONTROL_INPUT_TYPE_TIME, DYNAMIC_FORM_CONTROL_INPUT_TYPE_URL, DYNAMIC_FORM_CONTROL_INPUT_TYPE_WEEK, DYNAMIC_FORM_CONTROL_MAP_FN, DYNAMIC_FORM_CONTROL_TYPE_ARRAY, DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX, DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP, DYNAMIC_FORM_CONTROL_TYPE_COLORPICKER, DYNAMIC_FORM_CONTROL_TYPE_DATEPICKER, DYNAMIC_FORM_CONTROL_TYPE_EDITOR, DYNAMIC_FORM_CONTROL_TYPE_FILE_UPLOAD, DYNAMIC_FORM_CONTROL_TYPE_GROUP, DYNAMIC_FORM_CONTROL_TYPE_INPUT, DYNAMIC_FORM_CONTROL_TYPE_RADIO_GROUP, DYNAMIC_FORM_CONTROL_TYPE_RATING, DYNAMIC_FORM_CONTROL_TYPE_SELECT, DYNAMIC_FORM_CONTROL_TYPE_SLIDER, DYNAMIC_FORM_CONTROL_TYPE_SWITCH, DYNAMIC_FORM_CONTROL_TYPE_TEXTAREA, DYNAMIC_FORM_CONTROL_TYPE_TIMEPICKER, DYNAMIC_FORM_TEXTAREA_WRAP_HARD, DYNAMIC_FORM_TEXTAREA_WRAP_SOFT, DYNAMIC_MATCHERS, DYNAMIC_MATCHER_PROVIDERS, DYNAMIC_TEMPLATE_DIRECTIVE_ALIGNMENT, DYNAMIC_VALIDATORS, DisabledMatcher, DynamicCheckControlModel, DynamicCheckboxGroupModel, DynamicCheckboxModel, DynamicColorPickerModel, DynamicDateControlModel, DynamicDatePickerModel, DynamicEditorModel, DynamicFileControlModel, DynamicFileUploadModel, DynamicFormArrayComponent, DynamicFormArrayGroupModel, DynamicFormArrayModel, DynamicFormComponent, DynamicFormComponentService, DynamicFormControlComponent, DynamicFormControlContainerComponent, DynamicFormControlEventType, DynamicFormControlModel, DynamicFormGroupModel, DynamicFormLayoutService, DynamicFormOption, DynamicFormRelationService, DynamicFormService, DynamicFormValidationService, DynamicFormValueControlModel, DynamicFormsCoreModule, DynamicIdDirective, DynamicInputControlModel, DynamicInputModel, DynamicListDirective, DynamicOptionControlModel, DynamicRadioGroupModel, DynamicRatingModel, DynamicSelectModel, DynamicSliderModel, DynamicSwitchModel, DynamicTemplateDirective, DynamicTemplateableFormControlComponent, DynamicTextAreaModel, DynamicTimePickerModel, HIDDEN_MATCHER, HiddenMatcher, MATCH_DISABLED, MATCH_ENABLED, MATCH_HIDDEN, MATCH_OPTIONAL, MATCH_REQUIRED, MATCH_VISIBLE, METADATA_KEY_SERIALIZABLE, OR_OPERATOR, REQUIRED_MATCHER, RequiredMatcher, getSerializables, isAddressToken, isBoolean, isContactField, isContactToken, isDynamicFormControlEvent, isField, isFunction, isNumber, isObject, isSectionToken, isString, maskFromString, maskToString, parseReviver, serializable, serialize, validate };
