// @ng-dynamic-forms/core/esm5/lib/model/form-group/dynamic-form-group.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DynamicFormControlModel } from "../dynamic-form-control.model";
import { serializable } from "../../decorator/serializable.decorator";
/** @type {?} */
export var DYNAMIC_FORM_CONTROL_TYPE_GROUP = "GROUP";
/**
 * @record
 */
export function DynamicFormGroupModelConfig() { }
if (false) {
    /** @type {?|undefined} */
    DynamicFormGroupModelConfig.prototype.group;
    /** @type {?|undefined} */
    DynamicFormGroupModelConfig.prototype.legend;
}
var DynamicFormGroupModel = /** @class */ (function (_super) {
    tslib_1.__extends(DynamicFormGroupModel, _super);
    function DynamicFormGroupModel(config, layout) {
        var _this = _super.call(this, config, layout) || this;
        _this.group = [];
        _this.type = DYNAMIC_FORM_CONTROL_TYPE_GROUP;
        _this.group = Array.isArray(config.group) ? config.group : [];
        _this.legend = config.legend || null;
        return _this;
    }
    /**
     * @param {?} index
     * @return {?}
     */
    DynamicFormGroupModel.prototype.get = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        return this.group[index];
    };
    /**
     * @param {?} index
     * @param {?} controlModel
     * @return {?}
     */
    DynamicFormGroupModel.prototype.set = /**
     * @param {?} index
     * @param {?} controlModel
     * @return {?}
     */
    function (index, controlModel) {
        this.group[index] = controlModel;
    };
    /**
     * @param {?} controlModel
     * @return {?}
     */
    DynamicFormGroupModel.prototype.add = /**
     * @param {?} controlModel
     * @return {?}
     */
    function (controlModel) {
        this.group.push(controlModel);
    };
    /**
     * @param {?} index
     * @param {?} controlModel
     * @return {?}
     */
    DynamicFormGroupModel.prototype.insert = /**
     * @param {?} index
     * @param {?} controlModel
     * @return {?}
     */
    function (index, controlModel) {
        this.group.splice(index, 0, controlModel);
    };
    /**
     * @param {?} index
     * @param {?} step
     * @return {?}
     */
    DynamicFormGroupModel.prototype.move = /**
     * @param {?} index
     * @param {?} step
     * @return {?}
     */
    function (index, step) {
        var _a;
        (_a = this.group).splice.apply(_a, tslib_1.__spread([index + step, 0], this.group.splice(index, 1)));
    };
    /**
     * @param {?} index
     * @return {?}
     */
    DynamicFormGroupModel.prototype.remove = /**
     * @param {?} index
     * @return {?}
     */
    function (index) {
        this.group.splice(index, 1);
    };
    /**
     * @return {?}
     */
    DynamicFormGroupModel.prototype.size = /**
     * @return {?}
     */
    function () {
        return this.group.length;
    };
    tslib_1.__decorate([
        serializable(),
        tslib_1.__metadata("design:type", Array)
    ], DynamicFormGroupModel.prototype, "group", void 0);
    tslib_1.__decorate([
        serializable(),
        tslib_1.__metadata("design:type", String)
    ], DynamicFormGroupModel.prototype, "legend", void 0);
    tslib_1.__decorate([
        serializable(),
        tslib_1.__metadata("design:type", String)
    ], DynamicFormGroupModel.prototype, "type", void 0);
    return DynamicFormGroupModel;
}(DynamicFormControlModel));
export { DynamicFormGroupModel };
if (false) {
    /** @type {?} */
    DynamicFormGroupModel.prototype.group;
    /** @type {?} */
    DynamicFormGroupModel.prototype.legend;
    /** @type {?} */
    DynamicFormGroupModel.prototype.type;
}
