// @ng-dynamic-forms/core/esm2015/lib/model/form-array/dynamic-form-array.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DynamicFormControlModel } from "../dynamic-form-control.model";
import { serializable, serialize } from "../../decorator/serializable.decorator";
import { isFunction, isNumber } from "../../utils/core.utils";
export class DynamicFormArrayGroupModel {
    /**
     * @param {?} context
     * @param {?=} group
     * @param {?=} index
     */
    constructor(context, group = [], index = -1) {
        this.$implicit = this;
        this.context = context;
        this.group = group;
        this.index = index;
    }
    /**
     * @return {?}
     */
    get parent() {
        return this.context;
    }
    /**
     * @param {?} index
     * @return {?}
     */
    get(index) {
        return this.group[index];
    }
    /**
     * @return {?}
     */
    toJSON() {
        return serialize(this);
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Array)
], DynamicFormArrayGroupModel.prototype, "group", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Number)
], DynamicFormArrayGroupModel.prototype, "index", void 0);
if (false) {
    /** @type {?} */
    DynamicFormArrayGroupModel.prototype.$implicit;
    /** @type {?} */
    DynamicFormArrayGroupModel.prototype.context;
    /** @type {?} */
    DynamicFormArrayGroupModel.prototype.group;
    /** @type {?} */
    DynamicFormArrayGroupModel.prototype.index;
}
/** @type {?} */
export const DYNAMIC_FORM_CONTROL_TYPE_ARRAY = "ARRAY";
/**
 * @record
 */
export function DynamicFormArrayModelConfig() { }
if (false) {
    /** @type {?|undefined} */
    DynamicFormArrayModelConfig.prototype.groupAsyncValidators;
    /** @type {?|undefined} */
    DynamicFormArrayModelConfig.prototype.groupFactory;
    /** @type {?|undefined} */
    DynamicFormArrayModelConfig.prototype.groupValidators;
    /** @type {?|undefined} */
    DynamicFormArrayModelConfig.prototype.groups;
    /** @type {?|undefined} */
    DynamicFormArrayModelConfig.prototype.initialCount;
}
export class DynamicFormArrayModel extends DynamicFormControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.groups = [];
        // only to recreate model from JSON
        this.type = DYNAMIC_FORM_CONTROL_TYPE_ARRAY;
        if (isFunction(config.groupFactory)) {
            this.groupFactory = config.groupFactory;
        }
        else {
            throw new Error("group factory function must be specified for DynamicFormArrayModel");
        }
        this.groupAsyncValidators = config.groupAsyncValidators || null;
        this.groupPrototype = this.groupFactory();
        this.groupValidators = config.groupValidators || null;
        this.initialCount = isNumber(config.initialCount) ? config.initialCount : 1;
        if (Array.isArray(config.groups)) {
            config.groups.forEach((/**
             * @param {?} arrayGroup
             * @param {?} index
             * @return {?}
             */
            (arrayGroup, index) => {
                this.groups.push(new DynamicFormArrayGroupModel(this, arrayGroup.group, arrayGroup.index || index));
            }));
        }
        else {
            for (let index = 0; index < this.initialCount; index++) {
                this.addGroup();
            }
        }
    }
    /**
     * @private
     * @return {?}
     */
    updateGroupIndex() {
        this.groups.forEach((/**
         * @param {?} group
         * @param {?} index
         * @return {?}
         */
        (group, index) => group.index = index));
    }
    /**
     * @return {?}
     */
    get size() {
        return this.groups.length;
    }
    /**
     * @param {?} index
     * @return {?}
     */
    get(index) {
        return this.groups[index];
    }
    /**
     * @return {?}
     */
    addGroup() {
        return this.insertGroup(this.groups.length);
    }
    /**
     * @param {?} index
     * @return {?}
     */
    insertGroup(index) {
        /** @type {?} */
        let group = new DynamicFormArrayGroupModel(this, this.groupFactory());
        this.groups.splice(index, 0, group);
        this.updateGroupIndex();
        return group;
    }
    /**
     * @param {?} index
     * @param {?} step
     * @return {?}
     */
    moveGroup(index, step) {
        this.groups.splice(index + step, 0, ...this.groups.splice(index, 1));
        this.updateGroupIndex();
    }
    /**
     * @param {?} index
     * @return {?}
     */
    removeGroup(index) {
        this.groups.splice(index, 1);
        this.updateGroupIndex();
    }
    /**
     * @return {?}
     */
    clear() {
        this.groups.splice(0);
        this.updateGroupIndex();
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Object)
], DynamicFormArrayModel.prototype, "groupAsyncValidators", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Object)
], DynamicFormArrayModel.prototype, "groupValidators", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Array)
], DynamicFormArrayModel.prototype, "groups", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Number)
], DynamicFormArrayModel.prototype, "initialCount", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Array)
], DynamicFormArrayModel.prototype, "groupPrototype", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicFormArrayModel.prototype, "type", void 0);
if (false) {
    /** @type {?} */
    DynamicFormArrayModel.prototype.groupAsyncValidators;
    /** @type {?} */
    DynamicFormArrayModel.prototype.groupFactory;
    /** @type {?} */
    DynamicFormArrayModel.prototype.groupValidators;
    /** @type {?} */
    DynamicFormArrayModel.prototype.groups;
    /** @type {?} */
    DynamicFormArrayModel.prototype.initialCount;
    /** @type {?} */
    DynamicFormArrayModel.prototype.groupPrototype;
    /** @type {?} */
    DynamicFormArrayModel.prototype.type;
}
