// @ng-dynamic-forms/core/esm2015/lib/model/dynamic-date-control.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DynamicFormValueControlModel } from "./dynamic-form-value-control.model";
import { serializable } from "../decorator/serializable.decorator";
/**
 * @record
 */
export function DynamicDateControlModelConfig() { }
if (false) {
    /** @type {?|undefined} */
    DynamicDateControlModelConfig.prototype.format;
    /** @type {?|undefined} */
    DynamicDateControlModelConfig.prototype.max;
    /** @type {?|undefined} */
    DynamicDateControlModelConfig.prototype.min;
    /** @type {?|undefined} */
    DynamicDateControlModelConfig.prototype.placeholder;
}
/**
 * @abstract
 */
export class DynamicDateControlModel extends DynamicFormValueControlModel {
    /**
     * @protected
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.format = config.format || null;
        this.max = config.max || null;
        this.min = config.min || null;
        this.placeholder = config.placeholder || null;
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicDateControlModel.prototype, "format", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Object)
], DynamicDateControlModel.prototype, "max", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Object)
], DynamicDateControlModel.prototype, "min", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicDateControlModel.prototype, "placeholder", void 0);
if (false) {
    /** @type {?} */
    DynamicDateControlModel.prototype.format;
    /** @type {?} */
    DynamicDateControlModel.prototype.max;
    /** @type {?} */
    DynamicDateControlModel.prototype.min;
    /** @type {?} */
    DynamicDateControlModel.prototype.placeholder;
}
