// @ng-dynamic-forms/core/esm2015/lib/model/dynamic-form-value-control.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { BehaviorSubject } from "rxjs";
import { DynamicFormControlModel } from "./dynamic-form-control.model";
import { serializable } from "../decorator/serializable.decorator";
import { isBoolean, isObject } from "../utils/core.utils";
/**
 * @record
 * @template T
 */
export function DynamicFormValueControlModelConfig() { }
if (false) {
    /** @type {?|undefined} */
    DynamicFormValueControlModelConfig.prototype.additional;
    /** @type {?|undefined} */
    DynamicFormValueControlModelConfig.prototype.hint;
    /** @type {?|undefined} */
    DynamicFormValueControlModelConfig.prototype.required;
    /** @type {?|undefined} */
    DynamicFormValueControlModelConfig.prototype.tabIndex;
    /** @type {?|undefined} */
    DynamicFormValueControlModelConfig.prototype.value;
}
/**
 * @abstract
 * @template T
 */
export class DynamicFormValueControlModel extends DynamicFormControlModel {
    /**
     * @protected
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.additional = isObject(config.additional) ? config.additional : null;
        this.hint = config.hint || null;
        this.required = isBoolean(config.required) ? config.required : false;
        this.tabIndex = config.tabIndex || null;
        this.value$ = new BehaviorSubject(config.value !== null && config.value !== undefined ? config.value : null);
        this.value$.subscribe((/**
         * @param {?} value
         * @return {?}
         */
        value => this._value = value));
        this.valueChanges = this.value$.asObservable();
    }
    /**
     * @return {?}
     */
    get value() {
        return this.value$.getValue();
    }
    /**
     * @param {?} value
     * @return {?}
     */
    set value(value) {
        this.value$.next(value);
    }
    /**
     * @param {?} key
     * @param {?=} defaultValue
     * @return {?}
     */
    getAdditional(key, defaultValue = undefined) {
        return this.additional !== null && this.additional.hasOwnProperty(key) ? this.additional[key] : defaultValue;
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Object)
], DynamicFormValueControlModel.prototype, "additional", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicFormValueControlModel.prototype, "hint", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Boolean)
], DynamicFormValueControlModel.prototype, "required", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Number)
], DynamicFormValueControlModel.prototype, "tabIndex", void 0);
tslib_1.__decorate([
    serializable("value"),
    tslib_1.__metadata("design:type", Object)
], DynamicFormValueControlModel.prototype, "_value", void 0);
if (false) {
    /** @type {?} */
    DynamicFormValueControlModel.prototype.additional;
    /** @type {?} */
    DynamicFormValueControlModel.prototype.hint;
    /** @type {?} */
    DynamicFormValueControlModel.prototype.required;
    /** @type {?} */
    DynamicFormValueControlModel.prototype.tabIndex;
    /**
     * @type {?}
     * @private
     */
    DynamicFormValueControlModel.prototype._value;
    /**
     * @type {?}
     * @private
     */
    DynamicFormValueControlModel.prototype.value$;
    /** @type {?} */
    DynamicFormValueControlModel.prototype.valueChanges;
}
