// @ng-dynamic-forms/core/esm2015/lib/model/file-upload/dynamic-file-upload.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DynamicFileControlModel } from "../dynamic-file-control.model";
import { serializable } from "../../decorator/serializable.decorator";
import { isBoolean, isNumber } from "../../utils/core.utils";
/** @type {?} */
export const DYNAMIC_FORM_CONTROL_TYPE_FILE_UPLOAD = "FILE_UPLOAD";
/**
 * @record
 */
export function DynamicFileUploadModelConfig() { }
if (false) {
    /** @type {?|undefined} */
    DynamicFileUploadModelConfig.prototype.accept;
    /** @type {?|undefined} */
    DynamicFileUploadModelConfig.prototype.autoUpload;
    /** @type {?|undefined} */
    DynamicFileUploadModelConfig.prototype.maxSize;
    /** @type {?|undefined} */
    DynamicFileUploadModelConfig.prototype.minSize;
    /** @type {?|undefined} */
    DynamicFileUploadModelConfig.prototype.removeUrl;
    /** @type {?|undefined} */
    DynamicFileUploadModelConfig.prototype.showFileList;
    /** @type {?|undefined} */
    DynamicFileUploadModelConfig.prototype.url;
}
export class DynamicFileUploadModel extends DynamicFileControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_FILE_UPLOAD;
        this.accept = Array.isArray(config.accept) ? config.accept : null;
        this.autoUpload = isBoolean(config.autoUpload) ? config.autoUpload : true;
        this.maxSize = isNumber(config.maxSize) ? config.maxSize : null;
        this.minSize = isNumber(config.minSize) ? config.minSize : null;
        this.removeUrl = config.removeUrl || null;
        this.showFileList = isBoolean(config.showFileList) ? config.showFileList : true;
        this.url = config.url || null;
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Array)
], DynamicFileUploadModel.prototype, "accept", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Boolean)
], DynamicFileUploadModel.prototype, "autoUpload", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Number)
], DynamicFileUploadModel.prototype, "maxSize", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Number)
], DynamicFileUploadModel.prototype, "minSize", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicFileUploadModel.prototype, "removeUrl", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Boolean)
], DynamicFileUploadModel.prototype, "showFileList", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicFileUploadModel.prototype, "url", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicFileUploadModel.prototype, "type", void 0);
if (false) {
    /** @type {?} */
    DynamicFileUploadModel.prototype.accept;
    /** @type {?} */
    DynamicFileUploadModel.prototype.autoUpload;
    /** @type {?} */
    DynamicFileUploadModel.prototype.maxSize;
    /** @type {?} */
    DynamicFileUploadModel.prototype.minSize;
    /** @type {?} */
    DynamicFileUploadModel.prototype.removeUrl;
    /** @type {?} */
    DynamicFileUploadModel.prototype.showFileList;
    /** @type {?} */
    DynamicFileUploadModel.prototype.url;
    /** @type {?} */
    DynamicFileUploadModel.prototype.type;
}
