// @ng-dynamic-forms/core/esm2015/lib/model/slider/dynamic-slider.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DynamicFormValueControlModel } from "../dynamic-form-value-control.model";
import { serializable } from "../../decorator/serializable.decorator";
import { isBoolean, isNumber } from "../../utils/core.utils";
/** @type {?} */
export const DYNAMIC_FORM_CONTROL_TYPE_SLIDER = "SLIDER";
/**
 * @record
 */
export function DynamicSliderModelConfig() { }
if (false) {
    /** @type {?|undefined} */
    DynamicSliderModelConfig.prototype.max;
    /** @type {?|undefined} */
    DynamicSliderModelConfig.prototype.min;
    /** @type {?|undefined} */
    DynamicSliderModelConfig.prototype.step;
    /** @type {?|undefined} */
    DynamicSliderModelConfig.prototype.vertical;
}
export class DynamicSliderModel extends DynamicFormValueControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_SLIDER;
        this.max = isNumber(config.max) ? config.max : 10;
        this.min = isNumber(config.min) ? config.min : 0;
        this.step = isNumber(config.step) ? config.step : 1;
        this.vertical = isBoolean(config.vertical) ? config.vertical : false;
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Number)
], DynamicSliderModel.prototype, "max", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Number)
], DynamicSliderModel.prototype, "min", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Number)
], DynamicSliderModel.prototype, "step", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Boolean)
], DynamicSliderModel.prototype, "vertical", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicSliderModel.prototype, "type", void 0);
if (false) {
    /** @type {?} */
    DynamicSliderModel.prototype.max;
    /** @type {?} */
    DynamicSliderModel.prototype.min;
    /** @type {?} */
    DynamicSliderModel.prototype.step;
    /** @type {?} */
    DynamicSliderModel.prototype.vertical;
    /** @type {?} */
    DynamicSliderModel.prototype.type;
}
