// @ng-dynamic-forms/core/esm2015/lib/service/dynamic-form-relation.matchers.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { InjectionToken } from "@angular/core";
import { DynamicFormValidationService } from "./dynamic-form-validation.service";
import { isObject } from "../utils/core.utils";
/** @type {?} */
export const MATCH_DISABLED = "DISABLED";
/** @type {?} */
export const MATCH_ENABLED = "ENABLED";
/** @type {?} */
export const MATCH_HIDDEN = "HIDDEN";
/** @type {?} */
export const MATCH_OPTIONAL = "OPTIONAL";
/** @type {?} */
export const MATCH_REQUIRED = "REQUIRED";
/** @type {?} */
export const MATCH_VISIBLE = "VISIBLE";
/** @type {?} */
export const AND_OPERATOR = "AND";
/** @type {?} */
export const OR_OPERATOR = "OR";
/**
 * @record
 */
export function DynamicFormControlMatcher() { }
if (false) {
    /** @type {?} */
    DynamicFormControlMatcher.prototype.match;
    /** @type {?} */
    DynamicFormControlMatcher.prototype.opposingMatch;
    /**
     * @param {?} hasMatch
     * @param {?} model
     * @param {?} control
     * @param {?} injector
     * @return {?}
     */
    DynamicFormControlMatcher.prototype.onChange = function (hasMatch, model, control, injector) { };
}
/** @type {?} */
export const DYNAMIC_MATCHERS = new InjectionToken("DYNAMIC_MATCHERS");
/** @type {?} */
export const DisabledMatcher = {
    match: MATCH_DISABLED,
    opposingMatch: MATCH_ENABLED,
    /**
     * @param {?} hasMatch
     * @param {?} model
     * @return {?}
     */
    onChange(hasMatch, model) {
        model.disabled = hasMatch;
    }
};
/** @type {?} */
export const HiddenMatcher = {
    match: MATCH_HIDDEN,
    opposingMatch: MATCH_VISIBLE,
    /**
     * @param {?} hasMatch
     * @param {?} model
     * @return {?}
     */
    onChange(hasMatch, model) {
        model.hidden = hasMatch;
    }
};
/** @type {?} */
export const RequiredMatcher = {
    match: MATCH_REQUIRED,
    opposingMatch: MATCH_OPTIONAL,
    /**
     * @param {?} hasMatch
     * @param {?} model
     * @param {?} control
     * @param {?} injector
     * @return {?}
     */
    onChange(hasMatch, model, control, injector) {
        /** @type {?} */
        let validatorsConfig = null;
        if (hasMatch) {
            validatorsConfig = isObject(model.validators) ? Object.assign({}, model.validators, { required: null }) : { required: null };
        }
        else {
            if (isObject(model.validators)) {
                delete model.validators["required"];
                validatorsConfig = Object.assign({}, model.validators);
            }
        }
        injector.get(DynamicFormValidationService).updateValidators(validatorsConfig, control, model);
    }
};
/** @type {?} */
export const DISABLED_MATCHER = {
    provide: DYNAMIC_MATCHERS,
    useValue: DisabledMatcher,
    multi: true
};
/** @type {?} */
export const HIDDEN_MATCHER = {
    provide: DYNAMIC_MATCHERS,
    useValue: HiddenMatcher,
    multi: true
};
/** @type {?} */
export const REQUIRED_MATCHER = {
    provide: DYNAMIC_MATCHERS,
    useValue: RequiredMatcher,
    multi: true
};
/** @type {?} */
export const DYNAMIC_MATCHER_PROVIDERS = [DISABLED_MATCHER, HIDDEN_MATCHER, REQUIRED_MATCHER];
