// @ng-dynamic-forms/core/esm2015/lib/model/colorpicker/dynamic-colorpicker.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DynamicFormValueControlModel } from "../dynamic-form-value-control.model";
import { serializable } from "../../decorator/serializable.decorator";
import { isBoolean, isString } from "../../utils/core.utils";
/** @type {?} */
export const DYNAMIC_FORM_CONTROL_TYPE_COLORPICKER = "COLORPICKER";
/**
 * @record
 */
export function DynamicColorPickerModelConfig() { }
if (false) {
    /** @type {?|undefined} */
    DynamicColorPickerModelConfig.prototype.format;
    /** @type {?|undefined} */
    DynamicColorPickerModelConfig.prototype.inline;
}
export class DynamicColorPickerModel extends DynamicFormValueControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_COLORPICKER;
        this.format = isString(config.format) ? config.format : null;
        this.inline = isBoolean(config.inline) ? config.inline : false;
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicColorPickerModel.prototype, "format", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Boolean)
], DynamicColorPickerModel.prototype, "inline", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicColorPickerModel.prototype, "type", void 0);
if (false) {
    /** @type {?} */
    DynamicColorPickerModel.prototype.format;
    /** @type {?} */
    DynamicColorPickerModel.prototype.inline;
    /** @type {?} */
    DynamicColorPickerModel.prototype.type;
}
