// @ng-dynamic-forms/core/esm2015/lib/model/switch/dynamic-switch.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DynamicCheckControlModel } from "../dynamic-check-control.model";
import { serializable } from "../../decorator/serializable.decorator";
/** @type {?} */
export const DYNAMIC_FORM_CONTROL_TYPE_SWITCH = "SWITCH";
/**
 * @record
 */
export function DynamicSwitchModelConfig() { }
if (false) {
    /** @type {?|undefined} */
    DynamicSwitchModelConfig.prototype.offLabel;
    /** @type {?|undefined} */
    DynamicSwitchModelConfig.prototype.onLabel;
}
export class DynamicSwitchModel extends DynamicCheckControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_SWITCH;
        this.offLabel = config.offLabel || null;
        this.onLabel = config.onLabel || null;
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicSwitchModel.prototype, "offLabel", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicSwitchModel.prototype, "onLabel", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicSwitchModel.prototype, "type", void 0);
if (false) {
    /** @type {?} */
    DynamicSwitchModel.prototype.offLabel;
    /** @type {?} */
    DynamicSwitchModel.prototype.onLabel;
    /** @type {?} */
    DynamicSwitchModel.prototype.type;
}
