// @ng-dynamic-forms/core/esm2015/lib/model/timepicker/dynamic-timepicker.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DynamicDateControlModel } from "../dynamic-date-control.model";
import { serializable } from "../../decorator/serializable.decorator";
import { isBoolean } from "../../utils/core.utils";
/** @type {?} */
export const DYNAMIC_FORM_CONTROL_TYPE_TIMEPICKER = "TIMEPICKER";
/**
 * @record
 */
export function DynamicTimePickerModelConfig() { }
if (false) {
    /** @type {?|undefined} */
    DynamicTimePickerModelConfig.prototype.meridian;
    /** @type {?|undefined} */
    DynamicTimePickerModelConfig.prototype.showSeconds;
}
export class DynamicTimePickerModel extends DynamicDateControlModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_TIMEPICKER;
        this.meridian = isBoolean(config.meridian) ? config.meridian : false;
        this.showSeconds = isBoolean(config.showSeconds) ? config.showSeconds : false;
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Boolean)
], DynamicTimePickerModel.prototype, "meridian", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Boolean)
], DynamicTimePickerModel.prototype, "showSeconds", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicTimePickerModel.prototype, "type", void 0);
if (false) {
    /** @type {?} */
    DynamicTimePickerModel.prototype.meridian;
    /** @type {?} */
    DynamicTimePickerModel.prototype.showSeconds;
    /** @type {?} */
    DynamicTimePickerModel.prototype.type;
}
