// @ng-dynamic-forms/core/esm2015/lib/model/checkbox/dynamic-checkbox-group.model.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import * as tslib_1 from "tslib";
import { DynamicFormGroupModel } from "../form-group/dynamic-form-group.model";
import { serializable } from "../../decorator/serializable.decorator";
/** @type {?} */
export const DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP = "CHECKBOX_GROUP";
export class DynamicCheckboxGroupModel extends DynamicFormGroupModel {
    /**
     * @param {?} config
     * @param {?=} layout
     */
    constructor(config, layout) {
        super(config, layout);
        this.type = DYNAMIC_FORM_CONTROL_TYPE_CHECKBOX_GROUP;
    }
    /**
     * @param {...?} indices
     * @return {?}
     */
    check(...indices) {
        indices.forEach((/**
         * @param {?} index
         * @return {?}
         */
        index => this.group[index].checked = true));
    }
    /**
     * @param {...?} indices
     * @return {?}
     */
    uncheck(...indices) {
        indices.forEach((/**
         * @param {?} index
         * @return {?}
         */
        index => this.group[index].checked = false));
    }
    /**
     * @return {?}
     */
    checkAll() {
        this.group.forEach((/**
         * @param {?} model
         * @return {?}
         */
        model => model.checked = true));
    }
    /**
     * @return {?}
     */
    uncheckAll() {
        this.group.forEach((/**
         * @param {?} model
         * @return {?}
         */
        model => model.checked = false));
    }
}
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", Array)
], DynamicCheckboxGroupModel.prototype, "group", void 0);
tslib_1.__decorate([
    serializable(),
    tslib_1.__metadata("design:type", String)
], DynamicCheckboxGroupModel.prototype, "type", void 0);
if (false) {
    /** @type {?} */
    DynamicCheckboxGroupModel.prototype.group;
    /** @type {?} */
    DynamicCheckboxGroupModel.prototype.type;
}
