// @ng-dynamic-forms/core/esm5/lib/service/dynamic-form-component.service.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { Inject, Injectable, InjectionToken, Optional } from "@angular/core";
import { isFunction, isNumber } from "../utils/core.utils";
import * as i0 from "@angular/core";
/** @type {?} */
export var DYNAMIC_FORM_CONTROL_MAP_FN = new InjectionToken("DYNAMIC_FORM_CONTROL_MAP_FN");
var DynamicFormComponentService = /** @class */ (function () {
    function DynamicFormComponentService(DYNAMIC_FORM_CONTROL_MAP_FN) {
        this.DYNAMIC_FORM_CONTROL_MAP_FN = DYNAMIC_FORM_CONTROL_MAP_FN;
        this.componentRefs = {};
        this.DYNAMIC_FORM_CONTROL_MAP_FN = (/** @type {?} */ (DYNAMIC_FORM_CONTROL_MAP_FN));
    }
    /**
     * @param {?} modelId
     * @param {?=} index
     * @return {?}
     */
    DynamicFormComponentService.prototype.getFormControlRef = /**
     * @param {?} modelId
     * @param {?=} index
     * @return {?}
     */
    function (modelId, index) {
        /** @type {?} */
        var ref = this.componentRefs[modelId];
        if (isNumber(index)) {
            return Array.isArray(ref) ? ref[index] : undefined;
        }
        else {
            return (/** @type {?} */ (ref));
        }
    };
    /**
     * @param {?} model
     * @param {?} instance
     * @param {?=} index
     * @return {?}
     */
    DynamicFormComponentService.prototype.registerFormControlRef = /**
     * @param {?} model
     * @param {?} instance
     * @param {?=} index
     * @return {?}
     */
    function (model, instance, index) {
        if (isNumber(index)) {
            /** @type {?} */
            var arrayRef = (/** @type {?} */ (this.componentRefs[model.id])) || [];
            if (Array.isArray(arrayRef)) {
                arrayRef.splice(index, 0, instance);
                this.componentRefs[model.id] = arrayRef;
            }
            else {
                console.warn("registerFormControlRef is called with index for a non-array form control: " + model.id);
            }
        }
        else {
            this.componentRefs[model.id] = instance;
        }
    };
    /**
     * @param {?} modelId
     * @param {?=} index
     * @return {?}
     */
    DynamicFormComponentService.prototype.unregisterFormControlRef = /**
     * @param {?} modelId
     * @param {?=} index
     * @return {?}
     */
    function (modelId, index) {
        /** @type {?} */
        var componentRef = this.componentRefs[modelId];
        if (isNumber(index)) {
            if (Array.isArray(componentRef) && componentRef[index] !== undefined) {
                componentRef.splice(index, 1);
            }
        }
        else if (componentRef !== undefined) {
            delete this.componentRefs[modelId];
        }
    };
    /**
     * @param {?} model
     * @return {?}
     */
    DynamicFormComponentService.prototype.getCustomComponentType = /**
     * @param {?} model
     * @return {?}
     */
    function (model) {
        return isFunction(this.DYNAMIC_FORM_CONTROL_MAP_FN) ? this.DYNAMIC_FORM_CONTROL_MAP_FN(model) : null;
    };
    DynamicFormComponentService.decorators = [
        { type: Injectable, args: [{
                    providedIn: "root"
                },] }
    ];
    /** @nocollapse */
    DynamicFormComponentService.ctorParameters = function () { return [
        { type: undefined, decorators: [{ type: Inject, args: [DYNAMIC_FORM_CONTROL_MAP_FN,] }, { type: Optional }] }
    ]; };
    /** @nocollapse */ DynamicFormComponentService.ngInjectableDef = i0.ɵɵdefineInjectable({ factory: function DynamicFormComponentService_Factory() { return new DynamicFormComponentService(i0.ɵɵinject(DYNAMIC_FORM_CONTROL_MAP_FN, 8)); }, token: DynamicFormComponentService, providedIn: "root" });
    return DynamicFormComponentService;
}());
export { DynamicFormComponentService };
if (false) {
    /**
     * @type {?}
     * @private
     */
    DynamicFormComponentService.prototype.componentRefs;
    /**
     * @type {?}
     * @private
     */
    DynamicFormComponentService.prototype.DYNAMIC_FORM_CONTROL_MAP_FN;
}
